

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Text Outputs &mdash; Tombo 1.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Plotting Commands" href="plotting.html" />
    <link rel="prev" title="Modified Base Detection" href="modified_base_detection.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Tombo
          

          
          </a>

          
            
            
              <div class="version">
                1.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tombo Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Tombo Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="resquiggle.html">Re-squiggle Algorithm</a></li>
<li class="toctree-l1"><a class="reference internal" href="modified_base_detection.html">Modified Base Detection</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Text Outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#tombo-text-output-browser-files"><code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">browser_files</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#motif-filtering-output">Motif Filtering Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tombo-text-output-signif-sequence-context"><code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">signif_sequence_context</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="plotting.html">Plotting Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="filtering.html">Read Filtering Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="rna.html">RNA Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_training.html">Model Training (Advanced Users Only)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tombo.html">tombo package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Tombo</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Text Outputs</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/text_output.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="text-outputs">
<h1>Text Outputs<a class="headerlink" href="#text-outputs" title="Permalink to this headline">¶</a></h1>
<p>Tombo provides two text outputs:</p>
<ol class="arabic simple">
<li>Genome Broser Files - Genome browser compatible per-genomic-base statistics</li>
<li>Fasta - Genomic sequence output surrounding identified modified base sites</li>
</ol>
<div class="section" id="tombo-text-output-browser-files">
<h2><code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">browser_files</span></code><a class="headerlink" href="#tombo-text-output-browser-files" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">browser_files</span></code> command takes in a set of reads (<code class="docutils literal notranslate"><span class="pre">--fast5-basedirs</span></code>) and/or a statistics file generated from a <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">detect_modifications</span></code> command (<code class="docutils literal notranslate"><span class="pre">--statistics-filename</span></code>). A control set of reads can also be provided (<code class="docutils literal notranslate"><span class="pre">--control-fast5-basedirs</span></code>). Output files will be produced for each requested statistic (both plus and minus strands) in <a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/wiggle.html">variableStep wiggle format</a> (or <a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/bedgraph.html">bedgraph format</a> for <code class="docutils literal notranslate"><span class="pre">--file-type</span> <span class="pre">coverage</span></code>).</p>
<p>Several statistics are available for output:</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">coverage</span></code> - The coverage level for mapped and validly re-squiggled reads</li>
<li><code class="docutils literal notranslate"><span class="pre">valid_coverage</span></code> - The coverage level for reads that are mapped, validly re-squiggled and outside the interval specified by <code class="docutils literal notranslate"><span class="pre">--single-read-threshold</span></code> specified in a <code class="docutils literal notranslate"><span class="pre">--statistics-filename</span></code>.</li>
<li><code class="docutils literal notranslate"><span class="pre">dampened_fraction</span></code> - The estimated fraction of significantly modified reads (non-<code class="docutils literal notranslate"><span class="pre">level_sample_compare</span></code> modified base detection methods only)</li>
<li><code class="docutils literal notranslate"><span class="pre">fraction</span></code> - The raw fraction of significantly modified reads (non-<code class="docutils literal notranslate"><span class="pre">level_sample_compare</span></code> modified base detection methods only)</li>
<li><code class="docutils literal notranslate"><span class="pre">statistic</span></code> - Statistic produced from <code class="docutils literal notranslate"><span class="pre">level_sample_compare</span></code> method</li>
<li><code class="docutils literal notranslate"><span class="pre">signal</span></code> - The mean signal level across all reads mapped to this location</li>
<li><code class="docutils literal notranslate"><span class="pre">signal_sd</span></code> - The mean signal standard deviation across all reads mapped to this location (not available unless <code class="docutils literal notranslate"><span class="pre">--include-event-stdev</span></code> was provided in <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">resquiggle</span></code> command)</li>
<li><code class="docutils literal notranslate"><span class="pre">dwell</span></code> - The mean number of raw observations observed assigned to this location</li>
<li><code class="docutils literal notranslate"><span class="pre">difference</span></code> - The difference in normalized signal level between a sample and control set of reads</li>
</ul>
<div class="admonition hint">
<p class="first admonition-title">Hint</p>
<p class="last">The <code class="docutils literal notranslate"><span class="pre">dampened_fraction</span></code> output adds psuedo-counts to the detected number of un-modified and modified reads at each tested location (as specified by the <code class="docutils literal notranslate"><span class="pre">--coverage-dampen-counts</span></code> option), while the <code class="docutils literal notranslate"><span class="pre">fraction</span></code> option returns the raw fraction of modified reads at any reference site from <code class="docutils literal notranslate"><span class="pre">detect_modifications</span></code> results. The <code class="docutils literal notranslate"><span class="pre">dampen_fraction</span></code> output is intended to allow the inclusion of low coverage regions in downstream analysis without causing potentially false positive site at the top of rank lists. Visualize different values of the <code class="docutils literal notranslate"><span class="pre">--coverage-dampen-counts</span></code> option with the included <code class="docutils literal notranslate"><span class="pre">scripts/test_beta_priors.R</span></code> script.</p>
</div>
<div class="section" id="motif-filtering-output">
<h3>Motif Filtering Output<a class="headerlink" href="#motif-filtering-output" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">browser_files</span></code> contains options <code class="docutils literal notranslate"><span class="pre">--motif-descriptions</span></code> and <code class="docutils literal notranslate"><span class="pre">--genome-fasta</span></code> enabling computed statistics output to be restrcited to only those locations at known/putative motif-centered modifications. These options apply to the <code class="docutils literal notranslate"><span class="pre">fraction</span></code>, <code class="docutils literal notranslate"><span class="pre">dampened_fraction</span></code> and <code class="docutils literal notranslate"><span class="pre">valid_coverage</span></code> file types.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><code class="docutils literal notranslate"><span class="pre">signal</span></code>, <code class="docutils literal notranslate"><span class="pre">signal_sd</span></code>, <code class="docutils literal notranslate"><span class="pre">dwell</span></code> and <code class="docutils literal notranslate"><span class="pre">difference</span></code> require each reads’ event level data to be extracted from the raw read files and thus may be quite slow. <code class="docutils literal notranslate"><span class="pre">valid_coverage</span></code>, <code class="docutils literal notranslate"><span class="pre">fraction</span></code> , <code class="docutils literal notranslate"><span class="pre">dampened_fraction</span></code> and <code class="docutils literal notranslate"><span class="pre">statistic</span></code> can be extracted from the tombo statistics files and <code class="docutils literal notranslate"><span class="pre">coverage</span></code> from the Tombo index, which is much faster.</p>
<p class="last">The <code class="docutils literal notranslate"><span class="pre">signal</span></code>, <code class="docutils literal notranslate"><span class="pre">signal_sd</span></code>, <code class="docutils literal notranslate"><span class="pre">dwell</span></code> and <code class="docutils literal notranslate"><span class="pre">difference</span></code> outputs all require the <code class="docutils literal notranslate"><span class="pre">--fast5-basedirs</span></code> option, the <code class="docutils literal notranslate"><span class="pre">valid_coverage</span></code>, <code class="docutils literal notranslate"><span class="pre">fraction</span></code> , <code class="docutils literal notranslate"><span class="pre">dampened_fraction</span></code> and <code class="docutils literal notranslate"><span class="pre">statistic</span></code> outputs require the <code class="docutils literal notranslate"><span class="pre">--statistics-filename</span></code> option, and <code class="docutils literal notranslate"><span class="pre">coverage</span></code> output requires one or the other.</p>
</div>
<p>Files will be output to individual wiggle files (two per statistic for plus and minus genomic strand) in the following format <code class="docutils literal notranslate"><span class="pre">[wiggle-basename].[wiggle-type].[sample|control]?.[plus|minus].wig</span></code></p>
</div>
</div>
<div class="section" id="tombo-text-output-signif-sequence-context">
<h2><code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">signif_sequence_context</span></code><a class="headerlink" href="#tombo-text-output-signif-sequence-context" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">signif_sequence_context</span></code> command writes the genome sequence surrounding unique genomic positions with the largest estimated fraction of modified bases. This can be useful for several tasks related to modified base detection including motif discovery.</p>
<p>To run <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">signif_sequence_context</span></code>, a <code class="docutils literal notranslate"><span class="pre">--statistics-filename</span></code> is required to extract the most significant locations and either a <code class="docutils literal notranslate"><span class="pre">--fast5-basedirs</span></code> or <code class="docutils literal notranslate"><span class="pre">--genome-fasta</span></code> is required to extract the genomic sequence. Several options are availble for selecting the sequence to be output:</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">--num-regions</span></code> - Defines the number of unique locations to be output</li>
<li><code class="docutils literal notranslate"><span class="pre">--num-bases</span></code> - Defines the number of bases to be output surrounding the significant locations</li>
</ul>
<p>The output of this command could be used to determine sequence contexts consistently modified within a sample. Example <a class="reference external" href="http://meme-suite.org/doc/meme.html">meme</a> command line modified base motif detection command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tombo detect_modifications de_novo --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
             --statistics-file-basename sample.de_novo
tombo text_output signif_sequence_context --statistics-filename sample.de_novo.tombo.stats <span class="se">\</span>
             --genome-fasta reference.fasta --num-regions <span class="m">1000</span> --num-bases <span class="m">50</span>
./meme -oc tombo.de_novo_motif_detection.meme -dna -mod zoops tombo_results.significant_regions.fasta
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="plotting.html" class="btn btn-neutral float-right" title="Plotting Commands" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="modified_base_detection.html" class="btn btn-neutral" title="Modified Base Detection" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-18, Oxford Nanopore Technologies

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script async="async" type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>