

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>tombo.tombo_helper &mdash; Tombo 1.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> Tombo
          

          
          </a>

          
            
            
              <div class="version">
                1.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials.html">Tombo Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Tombo Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../resquiggle.html">Re-squiggle Algorithm</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modified_base_detection.html">Modified Base Detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../text_output.html">Text Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../plotting.html">Plotting Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../filtering.html">Read Filtering Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../rna.html">RNA Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../model_training.html">Model Training (Advanced Users Only)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tombo.html">tombo package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Tombo</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>tombo.tombo_helper</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for tombo.tombo_helper</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">division</span><span class="p">,</span> <span class="n">unicode_literals</span><span class="p">,</span> <span class="n">absolute_import</span>

<span class="kn">from</span> <span class="nn">builtins</span> <span class="k">import</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">range</span><span class="p">,</span> <span class="nb">dict</span><span class="p">,</span> <span class="nb">map</span><span class="p">,</span> <span class="nb">zip</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">io</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">queue</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="c1"># Future warning from cython in h5py</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">FutureWarning</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">h5py</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="nb">all</span><span class="o">=</span><span class="s1">&#39;raise&#39;</span><span class="p">)</span>

<span class="kn">from</span> <span class="nn">time</span> <span class="k">import</span> <span class="n">sleep</span>
<span class="kn">from</span> <span class="nn">time</span> <span class="k">import</span> <span class="n">strftime</span>
<span class="kn">from</span> <span class="nn">operator</span> <span class="k">import</span> <span class="n">itemgetter</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">repeat</span><span class="p">,</span> <span class="n">islice</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="k">import</span> <span class="n">Process</span><span class="p">,</span> <span class="n">Queue</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span><span class="p">,</span> <span class="n">namedtuple</span>

<span class="k">if</span> <span class="n">sys</span><span class="o">.</span><span class="n">version_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
    <span class="n">unicode</span> <span class="o">=</span> <span class="nb">str</span>

<span class="c1"># import tombo functions</span>
<span class="kn">from</span> <span class="nn">._version</span> <span class="k">import</span> <span class="n">TOMBO_VERSION</span>
<span class="kn">from</span> <span class="nn">._default_parameters</span> <span class="k">import</span> <span class="n">PHRED_BASE</span><span class="p">,</span> <span class="n">DNA_SAMP_TYPE</span><span class="p">,</span> <span class="n">RNA_SAMP_TYPE</span>
<span class="kn">from</span> <span class="nn">._c_helper</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">c_new_mean_stds</span><span class="p">,</span> <span class="n">c_new_means</span><span class="p">,</span> <span class="n">c_valid_cpts_w_cap</span><span class="p">,</span>
    <span class="n">c_valid_cpts_w_cap_t_test</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">._c_dynamic_programming</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">c_banded_traceback</span><span class="p">,</span> <span class="n">c_adaptive_banded_forward_pass</span><span class="p">)</span>


<span class="c1"># list of classes/functions to include in API</span>
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s1">&#39;readData&#39;</span><span class="p">,</span> <span class="s1">&#39;intervalData&#39;</span><span class="p">,</span> <span class="s1">&#39;TomboReads&#39;</span><span class="p">,</span>
    <span class="s1">&#39;TomboMotif&#39;</span><span class="p">,</span> <span class="s1">&#39;parse_motif_descs&#39;</span><span class="p">,</span>
    <span class="s1">&#39;resquiggleParams&#39;</span><span class="p">,</span> <span class="s1">&#39;startClipParams&#39;</span><span class="p">,</span>
    <span class="s1">&#39;resquiggleResults&#39;</span><span class="p">,</span> <span class="s1">&#39;alignInfo&#39;</span><span class="p">,</span> <span class="s1">&#39;genomeLocation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;sequenceData&#39;</span><span class="p">,</span> <span class="s1">&#39;dpResults&#39;</span><span class="p">,</span> <span class="s1">&#39;scaleValues&#39;</span><span class="p">,</span> <span class="s1">&#39;Fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;seqSampleType&#39;</span><span class="p">,</span>
    <span class="s1">&#39;get_seq_sample_type&#39;</span><span class="p">,</span> <span class="s1">&#39;get_raw_read_slot&#39;</span><span class="p">,</span>
    <span class="s1">&#39;get_single_slot_read_centric&#39;</span><span class="p">,</span> <span class="s1">&#39;get_multiple_slots_read_centric&#39;</span><span class="p">]</span>


<span class="n">VERBOSE</span> <span class="o">=</span> <span class="kc">True</span>

<span class="c1"># single base conversion for motifs</span>
<span class="n">SINGLE_LETTER_CODE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">:</span><span class="s1">&#39;[CGT]&#39;</span><span class="p">,</span>
    <span class="s1">&#39;D&#39;</span><span class="p">:</span><span class="s1">&#39;[AGT]&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">:</span><span class="s1">&#39;[ACT]&#39;</span><span class="p">,</span> <span class="s1">&#39;K&#39;</span><span class="p">:</span><span class="s1">&#39;[GT]&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;[AC]&#39;</span><span class="p">,</span>
    <span class="s1">&#39;N&#39;</span><span class="p">:</span><span class="s1">&#39;[ACGT]&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">:</span><span class="s1">&#39;[AG]&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">:</span><span class="s1">&#39;[CG]&#39;</span><span class="p">,</span> <span class="s1">&#39;V&#39;</span><span class="p">:</span><span class="s1">&#39;[ACG]&#39;</span><span class="p">,</span>
    <span class="s1">&#39;W&#39;</span><span class="p">:</span><span class="s1">&#39;[AT]&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span><span class="s1">&#39;[CT]&#39;</span><span class="p">}</span>
<span class="n">INVALID_BASES</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;[^ACGT]&#39;</span><span class="p">)</span>
<span class="n">INVALID_BASE_RUNS</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;[^ACGT]+&#39;</span><span class="p">)</span>


<span class="c1">###############################</span>
<span class="c1">###### Custom TomboError ######</span>
<span class="c1">###############################</span>

<span class="k">class</span> <span class="nc">TomboError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="k">pass</span>


<span class="c1">######################################</span>
<span class="c1">###### Cython Wrapper Functions ######</span>
<span class="c1">######################################</span>

<span class="c1"># can&#39;t import TomboError in cython so wrap these functions in python</span>
<span class="k">def</span> <span class="nf">valid_cpts_w_cap</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">valid_cpts</span> <span class="o">=</span> <span class="n">c_valid_cpts_w_cap</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">NotImplementedError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="n">unicode</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
    <span class="n">valid_cpts</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">valid_cpts</span>

<span class="k">def</span> <span class="nf">valid_cpts_w_cap_t_test</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">valid_cpts</span> <span class="o">=</span> <span class="n">c_valid_cpts_w_cap_t_test</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">NotImplementedError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="n">unicode</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
    <span class="n">valid_cpts</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">valid_cpts</span>

<span class="k">def</span> <span class="nf">banded_traceback</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">c_banded_traceback</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">NotImplementedError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="n">unicode</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="k">def</span> <span class="nf">adaptive_banded_forward_pass</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">c_adaptive_banded_forward_pass</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">NotImplementedError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="n">unicode</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>


<span class="c1">################################</span>
<span class="c1">###### Global Namedtuples ######</span>
<span class="c1">################################</span>

<div class="viewcode-block" id="alignInfo"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.alignInfo">[docs]</a><span class="k">class</span> <span class="nc">alignInfo</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;alignInfo&#39;</span><span class="p">,</span>
        <span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Subgroup&#39;</span><span class="p">,</span> <span class="s1">&#39;ClipStart&#39;</span><span class="p">,</span> <span class="s1">&#39;ClipEnd&#39;</span><span class="p">,</span>
         <span class="s1">&#39;Insertions&#39;</span><span class="p">,</span> <span class="s1">&#39;Deletions&#39;</span><span class="p">,</span> <span class="s1">&#39;Matches&#39;</span><span class="p">,</span> <span class="s1">&#39;Mismatches&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Information from genomic read alignment</span>

<span class="sd">    Args:</span>
<span class="sd">        ID (str): read identifier</span>
<span class="sd">        Subgroup (str): sub-group location containing read information (e.g. &#39;BaseCalled_template&#39;)</span>
<span class="sd">        ClipStart (int): number of bases clipped from beginning of basecalls</span>
<span class="sd">        ClipEnd (int): number of bases clipped from end of basecalls</span>
<span class="sd">        Insertions (int): number of inserted bases in alignment</span>
<span class="sd">        Deletions (int): number of delected bases in alignment</span>
<span class="sd">        Matches (int): number of matched bases in alignment</span>
<span class="sd">        Mismatches (int): number of mis-matched bases in alignment</span>
<span class="sd">    &quot;&quot;&quot;</span></div>

<span class="c1"># TODO convert rna to rev_sig</span>
<div class="viewcode-block" id="readData"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.readData">[docs]</a><span class="k">class</span> <span class="nc">readData</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span><span class="s1">&#39;readData&#39;</span><span class="p">,</span> <span class="p">(</span>
        <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;filtered&#39;</span><span class="p">,</span> <span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">,</span> <span class="s1">&#39;fn&#39;</span><span class="p">,</span>
        <span class="s1">&#39;corr_group&#39;</span><span class="p">,</span> <span class="s1">&#39;rna&#39;</span><span class="p">,</span> <span class="s1">&#39;sig_match_score&#39;</span><span class="p">,</span> <span class="s1">&#39;mean_q_score&#39;</span><span class="p">,</span> <span class="s1">&#39;read_id&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Nanopore read meta-data</span>

<span class="sd">    Example::</span>

<span class="sd">        r_data = tombo_helper.readData(</span>
<span class="sd">            start=0, end=1000, filtered=False, read_start_rel_to_raw=100,</span>
<span class="sd">            strand=&#39;+&#39;, fn=&#39;path/to/read.fast5&#39;,</span>
<span class="sd">            corr_group=&#39;RawGenomeCorrected_000&#39;, rna=False,</span>
<span class="sd">            sig_match_score=1.0, mean_q_score=10.0)</span>

<span class="sd">        # output is a list of readData objects</span>
<span class="sd">        reads_index = tombo_helper.TomboReads([&#39;test_data/native_reads&#39;,])</span>
<span class="sd">        cs_reads = reads_index.get_cs_reads(&#39;chr20&#39;, &#39;+&#39;)</span>

<span class="sd">    Args:</span>
<span class="sd">        start (int): 0-based start mapped position</span>
<span class="sd">        end (int): 0-based open interval end mapped position</span>
<span class="sd">        filtered (bool): is this read filtered?</span>
<span class="sd">        read_start_rel_to_raw (int): start (in raw signal vector) of assigned bases</span>
<span class="sd">        strand (str): read mapped strand (&#39;+&#39;, &#39;-&#39; or None)</span>
<span class="sd">        fn (str): raw read filename.</span>
<span class="sd">        corr_group (str): --corrected-group slot specified in re-squiggle command</span>
<span class="sd">        rna (bool): is this read RNA?</span>
<span class="sd">        sig_match_score (float): signal matching score. Default: None</span>
<span class="sd">        mean_q_score (float): mean basecalling q-score. Default: None</span>
<span class="sd">        read_id (str): read identifier Default: None</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
<span class="c1"># set default values for sig_match_score, q_score and read_id</span>
<span class="n">readData</span><span class="o">.</span><span class="fm">__new__</span><span class="o">.</span><span class="vm">__defaults__</span> <span class="o">=</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

<div class="viewcode-block" id="scaleValues"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.scaleValues">[docs]</a><span class="k">class</span> <span class="nc">scaleValues</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;scaleValues&#39;</span><span class="p">,</span> <span class="p">(</span>
            <span class="s1">&#39;shift&#39;</span><span class="p">,</span> <span class="s1">&#39;scale&#39;</span><span class="p">,</span> <span class="s1">&#39;lower_lim&#39;</span><span class="p">,</span> <span class="s1">&#39;upper_lim&#39;</span><span class="p">,</span> <span class="s1">&#39;outlier_thresh&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Signal normaliztion scaling parameters. For details see https://nanoporetech.github.io/tombo/resquiggle.html#signal-normalization</span>

<span class="sd">    Args:</span>
<span class="sd">        shift (float): shift scaling parameter</span>
<span class="sd">        scale (float): scale scaling parameter</span>
<span class="sd">        lower_lim (float): lower windsorizing threshold</span>
<span class="sd">        upper_lim (float): upper windsorizing threshold</span>
<span class="sd">        outlier_thresh (float): outlier threshold used to define `lower_lim` and `upper_lim`</span>
<span class="sd">    &quot;&quot;&quot;</span></div>

<div class="viewcode-block" id="resquiggleParams"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.resquiggleParams">[docs]</a><span class="k">class</span> <span class="nc">resquiggleParams</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;resquiggleParams&#39;</span><span class="p">,</span>
        <span class="p">(</span><span class="s1">&#39;match_evalue&#39;</span><span class="p">,</span> <span class="s1">&#39;skip_pen&#39;</span><span class="p">,</span> <span class="s1">&#39;bandwidth&#39;</span><span class="p">,</span> <span class="s1">&#39;max_half_z_score&#39;</span><span class="p">,</span>
         <span class="s1">&#39;running_stat_width&#39;</span><span class="p">,</span> <span class="s1">&#39;min_obs_per_base&#39;</span><span class="p">,</span> <span class="s1">&#39;raw_min_obs_per_base&#39;</span><span class="p">,</span>
         <span class="s1">&#39;mean_obs_per_event&#39;</span><span class="p">,</span> <span class="s1">&#39;z_shift&#39;</span><span class="p">,</span> <span class="s1">&#39;stay_pen&#39;</span><span class="p">,</span> <span class="s1">&#39;use_t_test_seg&#39;</span><span class="p">,</span>
         <span class="s1">&#39;band_bound_thresh&#39;</span><span class="p">,</span> <span class="s1">&#39;start_bw&#39;</span><span class="p">,</span> <span class="s1">&#39;start_save_bw&#39;</span><span class="p">,</span> <span class="s1">&#39;start_n_bases&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Re-squiggle parameters</span>

<span class="sd">    Args:</span>
<span class="sd">        match_evalue (float): expected value for matching event to sequence</span>
<span class="sd">        skip_pen (float): penalty for skipped sequence position</span>
<span class="sd">        bandwidth (int): adaptive bandwidth</span>
<span class="sd">        max_half_z_score (float): windsorize half z-scores above this value</span>
<span class="sd">        running_stat_width (int): running neighboring window width for segmentation scoring</span>
<span class="sd">        min_obs_per_base (int): minimum observations per genomic base</span>
<span class="sd">        mean_obs_per_event (int): mean number of raw obs. per event during segmentation</span>
<span class="sd">        z_shift (float): amount to shift z-scores for DP (derived from match_evalue)</span>
<span class="sd">        stay_pen (float): stay penalty for DP (derived from match_evalue)</span>
<span class="sd">        use_t_test_seg (bool): use t-test segmentation criterion (default: raw neighboring window difference)</span>
<span class="sd">        band_bound_thresh (int): bandwidth boundary threshold for determining if a read potentially left the adaptive bandwidth</span>
<span class="sd">        start_bw (int): bandwidth for read start identification</span>
<span class="sd">        start_save_bw (int): save bandwidth for read start identification (if start_bw_fails)</span>
<span class="sd">        start_n_bases (int): number of genomic bases to use for read start identification</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
<span class="c1"># set default values for start params</span>
<span class="n">resquiggleParams</span><span class="o">.</span><span class="fm">__new__</span><span class="o">.</span><span class="vm">__defaults__</span> <span class="o">=</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

<span class="k">class</span> <span class="nc">trimRnaParams</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;trimRnaParams&#39;</span><span class="p">,</span>
        <span class="p">(</span><span class="s1">&#39;moving_window_size&#39;</span><span class="p">,</span> <span class="s1">&#39;min_running_values&#39;</span><span class="p">,</span> <span class="s1">&#39;thresh_scale&#39;</span><span class="p">,</span>
         <span class="s1">&#39;max_raw_obs&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Parameters to trim RNA adapters.</span>
<span class="sd">    &quot;&quot;&quot;</span>

<span class="k">class</span> <span class="nc">stallParams</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;stallParams&#39;</span><span class="p">,</span>
        <span class="p">(</span><span class="s1">&#39;window_size&#39;</span><span class="p">,</span> <span class="s1">&#39;threshold&#39;</span><span class="p">,</span> <span class="s1">&#39;min_consecutive_obs&#39;</span><span class="p">,</span> <span class="s1">&#39;edge_buffer&#39;</span><span class="p">,</span>
         <span class="c1"># percentile stall method params</span>
         <span class="s1">&#39;lower_pctl&#39;</span><span class="p">,</span> <span class="s1">&#39;upper_pctl&#39;</span><span class="p">,</span>
         <span class="c1"># mean windows stall method params</span>
         <span class="s1">&#39;mini_window_size&#39;</span><span class="p">,</span> <span class="s1">&#39;n_windows&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Parameters to identify RNA stalls</span>
<span class="sd">    &quot;&quot;&quot;</span>
<span class="c1"># set method specific params to None</span>
<span class="n">stallParams</span><span class="o">.</span><span class="fm">__new__</span><span class="o">.</span><span class="vm">__defaults__</span> <span class="o">=</span> <span class="p">(</span><span class="kc">None</span><span class="p">,)</span> <span class="o">*</span> <span class="mi">4</span>

<div class="viewcode-block" id="startClipParams"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.startClipParams">[docs]</a><span class="k">class</span> <span class="nc">startClipParams</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;startClipParams&#39;</span><span class="p">,</span>
        <span class="p">(</span><span class="s1">&#39;bandwidth&#39;</span><span class="p">,</span> <span class="s1">&#39;num_genome_bases&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Parameters to identify read start using bases clipped from mapping</span>

<span class="sd">    Args:</span>
<span class="sd">        bandwidth (int): bandwidth</span>
<span class="sd">        num_genome_bases (int): number of genome bases</span>
<span class="sd">    &quot;&quot;&quot;</span></div>

<div class="viewcode-block" id="resquiggleResults"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.resquiggleResults">[docs]</a><span class="k">class</span> <span class="nc">resquiggleResults</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;resquiggleResults&#39;</span><span class="p">,</span>
        <span class="p">(</span><span class="s1">&#39;align_info&#39;</span><span class="p">,</span> <span class="s1">&#39;genome_loc&#39;</span><span class="p">,</span> <span class="s1">&#39;genome_seq&#39;</span><span class="p">,</span> <span class="s1">&#39;mean_q_score&#39;</span><span class="p">,</span>
         <span class="s1">&#39;raw_signal&#39;</span><span class="p">,</span> <span class="s1">&#39;channel_info&#39;</span><span class="p">,</span> <span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">,</span> <span class="s1">&#39;segs&#39;</span><span class="p">,</span>
         <span class="s1">&#39;scale_values&#39;</span><span class="p">,</span> <span class="s1">&#39;sig_match_score&#39;</span><span class="p">,</span> <span class="s1">&#39;norm_params_changed&#39;</span><span class="p">,</span>
         <span class="s1">&#39;start_clip_bases&#39;</span><span class="p">,</span> <span class="s1">&#39;stall_ints&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Re-squiggle results</span>

<span class="sd">    Args:</span>
<span class="sd">        align_info (:class:`tombo.tombo_helper.alignInfo`): read alignment information</span>
<span class="sd">        genome_loc (:class:`tombo.tombo_helper.genomeLocation`): genome mapping location</span>
<span class="sd">        genome_seq (str): mapped genome sequence</span>
<span class="sd">        mean_q_score (float): mean basecalling q-score</span>
<span class="sd">        raw_signal (np.array::np.float64): raw signal (i.e. un-segmented; signal may be normalized) (optional)</span>
<span class="sd">        channel_info (:class:`tombo.tombo_helper.channelInfo`): channel information (optional)</span>
<span class="sd">        read_start_rel_to_raw (int): read start within raw signal (optional)</span>
<span class="sd">        segs (np.array::np.int64): relative raw signal segment positions (optional)</span>
<span class="sd">        scale_values (:class:`tombo.tombo_helper.scaleValues`): signal normalization scale values (optional)</span>
<span class="sd">        sig_match_score (float): expected to observed signal score (see :class:`tombo.tombo_stats.get_read_seg_score`; optional)</span>
<span class="sd">        norm_params_changed (bool): were scale parameters updated (optional)</span>
<span class="sd">        start_clip_bases (str): mapping clipped bases from start of read (optional)</span>
<span class="sd">        stall_ints (list): list of rna stall locations within raw signal (optional)</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
<span class="c1"># set default None values for when just mapping results are included</span>
<span class="n">resquiggleResults</span><span class="o">.</span><span class="fm">__new__</span><span class="o">.</span><span class="vm">__defaults__</span> <span class="o">=</span> <span class="p">(</span><span class="kc">None</span><span class="p">,)</span> <span class="o">*</span> <span class="mi">9</span>

<div class="viewcode-block" id="dpResults"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.dpResults">[docs]</a><span class="k">class</span> <span class="nc">dpResults</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;dpResults&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">,</span> <span class="s1">&#39;segs&#39;</span><span class="p">,</span> <span class="s1">&#39;ref_means&#39;</span><span class="p">,</span> <span class="s1">&#39;ref_sds&#39;</span><span class="p">,</span>
                      <span class="s1">&#39;genome_seq&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Dynamic programming results</span>

<span class="sd">    Args:</span>
<span class="sd">        read_start_rel_to_raw (int): read start within raw signal</span>
<span class="sd">        segs (np.array::np.int64): relative raw signal segment positions</span>
<span class="sd">        ref_means (np.array::np.float64): expected signal levels</span>
<span class="sd">        ref_sds (np.array::np.float64): expected SD of signal levels</span>
<span class="sd">        genome_seq (str): mapped genome sequence</span>
<span class="sd">    &quot;&quot;&quot;</span></div>

<div class="viewcode-block" id="genomeLocation"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.genomeLocation">[docs]</a><span class="k">class</span> <span class="nc">genomeLocation</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span><span class="s1">&#39;genomeLocation&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;Start&#39;</span><span class="p">,</span> <span class="s1">&#39;Strand&#39;</span><span class="p">,</span> <span class="s1">&#39;Chrom&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Genomic location</span>

<span class="sd">    Args:</span>
<span class="sd">        Start (int): 0-based genomic start position</span>
<span class="sd">        Strand (str): Strand (should be &#39;+&#39; or &#39;-&#39;)</span>
<span class="sd">        Chrom (str): Chromosome</span>
<span class="sd">    &quot;&quot;&quot;</span></div>

<div class="viewcode-block" id="sequenceData"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.sequenceData">[docs]</a><span class="k">class</span> <span class="nc">sequenceData</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span><span class="s1">&#39;sequenceData&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;seq&#39;</span><span class="p">,</span> <span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;mean_q_score&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Read sequence data from FASTQ record</span>

<span class="sd">    Args:</span>
<span class="sd">        seq (str): read sequence</span>
<span class="sd">        id (str): read id (extracted from FASTQ record line)</span>
<span class="sd">        mean_q_score (float): mean q-score</span>
<span class="sd">    &quot;&quot;&quot;</span></div>

<span class="k">class</span> <span class="nc">channelInfo</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;channelInfo&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;offset&#39;</span><span class="p">,</span> <span class="s1">&#39;range&#39;</span><span class="p">,</span> <span class="s1">&#39;digitisation&#39;</span><span class="p">,</span>
                        <span class="s1">&#39;number&#39;</span><span class="p">,</span> <span class="s1">&#39;sampling_rate&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Read channel information</span>

<span class="sd">    Args:</span>
<span class="sd">        offset (float): offset parameter</span>
<span class="sd">        range (float): range parameter</span>
<span class="sd">        digitisation (float): digitisation parameter</span>
<span class="sd">        number (int): channel number</span>
<span class="sd">        sampling_rate (int): number of raw samples per second</span>
<span class="sd">    &quot;&quot;&quot;</span>

<span class="k">class</span> <span class="nc">regionStats</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;regionStats&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;reg_frac_standard_base&#39;</span><span class="p">,</span> <span class="s1">&#39;reg_poss&#39;</span><span class="p">,</span> <span class="s1">&#39;chrm&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">,</span>
                        <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;reg_cov&#39;</span><span class="p">,</span> <span class="s1">&#39;ctrl_cov&#39;</span><span class="p">,</span> <span class="s1">&#39;valid_cov&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Region statistics</span>

<span class="sd">    Args:</span>
<span class="sd">        reg_frac_standard_base (np.array::np.float64): fraction of standard bases</span>
<span class="sd">        reg_poss (np.array::np.int64): positions for reported fractions</span>
<span class="sd">        chrm (str): chromosome name</span>
<span class="sd">        strand (str): strand (should be &#39;+&#39; or &#39;-&#39;)</span>
<span class="sd">        start (int): 0-based region start</span>
<span class="sd">        reg_cov (np.array::np.int64): region read depth</span>
<span class="sd">        ctrl_cov (np.array::np.int64): region control sample read depth</span>
<span class="sd">        valid_cov (np.array::np.int64): region valid (tested) read depth</span>
<span class="sd">    &quot;&quot;&quot;</span>

<span class="k">class</span> <span class="nc">groupStats</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;groupStats&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;reg_stats&#39;</span><span class="p">,</span> <span class="s1">&#39;reg_poss&#39;</span><span class="p">,</span> <span class="s1">&#39;chrm&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;reg_cov&#39;</span><span class="p">,</span> <span class="s1">&#39;ctrl_cov&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Region statistics</span>

<span class="sd">    Args:</span>
<span class="sd">        reg_stats (np.array::np.float64): statistic for group comparison</span>
<span class="sd">        reg_poss (np.array::np.int64): positions for reported fractions</span>
<span class="sd">        chrm (str): chromosome name</span>
<span class="sd">        strand (str): strand (should be &#39;+&#39; or &#39;-&#39;)</span>
<span class="sd">        start (int): 0-based region start</span>
<span class="sd">        reg_cov (np.array::np.int64): region read depth</span>
<span class="sd">        ctrl_cov (np.array::np.int64): region control sample read depth</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="seqSampleType"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.seqSampleType">[docs]</a><span class="k">class</span> <span class="nc">seqSampleType</span><span class="p">(</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;seqSampleType&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;rev_sig&#39;</span><span class="p">))):</span>
    <span class="sd">&quot;&quot;&quot;Description of a sequencing sample type</span>

<span class="sd">    Args:</span>
<span class="sd">        name (str): name of the sequencing sample</span>
<span class="sd">        rev_sig (bool): Is the raw signal reversed (3&#39; to 5&#39;)</span>
<span class="sd">    &quot;&quot;&quot;</span></div>


<span class="c1">######################################</span>
<span class="c1">###### Various Helper Functions ######</span>
<span class="c1">######################################</span>

<span class="k">def</span> <span class="nf">status_message</span><span class="p">(</span><span class="n">message</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="n">pre_str</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">indent</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">pre_str</span> <span class="o">+</span> <span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;[%H:%M:%S] &#39;</span><span class="p">)</span> <span class="o">+</span> <span class="n">message</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
    <span class="k">return</span>

<span class="k">def</span> <span class="nf">warning_message</span><span class="p">(</span><span class="n">message</span><span class="p">):</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
        <span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">20</span> <span class="o">+</span> <span class="s1">&#39; WARNING &#39;</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">20</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span> <span class="o">+</span>
        <span class="n">message</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
    <span class="k">return</span>

<span class="k">def</span> <span class="nf">error_message_and_exit</span><span class="p">(</span><span class="n">message</span><span class="p">):</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
        <span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">20</span> <span class="o">+</span> <span class="s1">&#39; ERROR &#39;</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">20</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span> <span class="o">+</span>
        <span class="n">message</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">()</span>
    <span class="k">return</span>

<span class="k">def</span> <span class="nf">resolve_path</span><span class="p">(</span><span class="n">fn_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Helper function to resolve relative and linked paths that might</span>
<span class="sd">    give other packages problems.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">realpath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="n">fn_path</span><span class="p">))</span>

<span class="n">COMP_BASES</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">ord</span><span class="p">,</span> <span class="s1">&#39;ACGT&#39;</span><span class="p">),</span> <span class="nb">map</span><span class="p">(</span><span class="nb">ord</span><span class="p">,</span> <span class="s1">&#39;TGCA&#39;</span><span class="p">)))</span>
<span class="k">def</span> <span class="nf">comp_seq</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Complement DNA sequence</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMP_BASES</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">rev_comp</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reverse complement DNA sequence</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMP_BASES</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">invalid_seq</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="n">INVALID_BASES</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>

<span class="n">U_TO_T</span> <span class="o">=</span> <span class="p">{</span><span class="nb">ord</span><span class="p">(</span><span class="s1">&#39;U&#39;</span><span class="p">):</span><span class="nb">ord</span><span class="p">(</span><span class="s1">&#39;T&#39;</span><span class="p">)}</span>
<span class="k">def</span> <span class="nf">rev_transcribe</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Convert U bases to T</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">U_TO_T</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">get_mean_q_score</span><span class="p">(</span><span class="n">read_q</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">sys</span><span class="o">.</span><span class="n">version_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">q_val</span> <span class="o">-</span> <span class="n">PHRED_BASE</span>
                        <span class="k">for</span> <span class="n">q_val</span> <span class="ow">in</span> <span class="n">read_q</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s1">&#39;ASCII&#39;</span><span class="p">)])</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="nb">ord</span><span class="p">(</span><span class="n">q_val</span><span class="p">)</span> <span class="o">-</span> <span class="n">PHRED_BASE</span>
                    <span class="k">for</span> <span class="n">q_val</span> <span class="ow">in</span> <span class="n">read_q</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s1">&#39;ASCII&#39;</span><span class="p">)])</span>

<span class="k">def</span> <span class="nf">get_chrm_sizes</span><span class="p">(</span><span class="n">reads_index</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get covered chromosome sizes from a set of reads</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">strand_chrm_sizes</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_read_cov</span> <span class="ow">in</span> <span class="n">reads_index</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">strand_chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span>
                    <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_read_cov</span><span class="p">))</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">continue</span>
    <span class="k">if</span> <span class="n">ctrl_reads_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_read_cov</span> <span class="ow">in</span> <span class="n">ctrl_reads_index</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">strand_chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span>
                        <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_read_cov</span><span class="p">))</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">continue</span>

    <span class="n">chrm_sizes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strnd_sizes</span> <span class="ow">in</span> <span class="n">strand_chrm_sizes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">strnd_sizes</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">continue</span>

    <span class="k">return</span> <span class="n">chrm_sizes</span>

<span class="k">def</span> <span class="nf">parse_genome_locations</span><span class="p">(</span><span class="n">genome_locs</span><span class="p">,</span> <span class="n">default_strand</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parse genome location strings and convert to 0-based coordinates</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parsed_locs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">chrm_pos_strand</span> <span class="ow">in</span> <span class="n">genome_locs</span><span class="p">:</span>
        <span class="c1"># strip off any quotes and return up to the first 3 values</span>
        <span class="n">split_vals</span> <span class="o">=</span> <span class="n">chrm_pos_strand</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
            <span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[:</span><span class="mi">3</span><span class="p">]</span>
        <span class="c1"># default to plus strand if not specified</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">split_vals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">error_message_and_exit</span><span class="p">(</span>
                <span class="s1">&#39;Invalid genome location provided: &#39;</span> <span class="o">+</span> <span class="n">chrm_pos_strand</span> <span class="o">+</span>
                <span class="s1">&#39;</span><span class="se">\n\t\t</span><span class="s1">Try adding quotation marks around specified genome &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;locations (especially for sequence identifiers with &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;special characters).&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">split_vals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">parsed_locs</span><span class="o">.</span><span class="n">append</span><span class="p">((</span>
                <span class="n">split_vals</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">split_vals</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">default_strand</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">parsed_locs</span><span class="o">.</span><span class="n">append</span><span class="p">((</span>
                <span class="n">split_vals</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">split_vals</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">split_vals</span><span class="p">[</span><span class="mi">2</span><span class="p">]))</span>

    <span class="k">return</span> <span class="n">parsed_locs</span>

<span class="k">def</span> <span class="nf">parse_genome_regions</span><span class="p">(</span><span class="n">all_regs_text</span><span class="p">):</span>
    <span class="n">parsed_regs</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">include_whole_chrms</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">reg_text</span> <span class="ow">in</span> <span class="n">all_regs_text</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">chrm_reg</span> <span class="o">=</span> <span class="n">reg_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">chrm_reg</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">chrm</span> <span class="o">=</span> <span class="n">chrm_reg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">reg_pos</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">include_whole_chrms</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">chrm</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">chrm_reg</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">chrm</span><span class="p">,</span> <span class="n">reg_pos</span> <span class="o">=</span> <span class="n">chrm_reg</span>
                <span class="n">reg_pos</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)),</span>
                                   <span class="n">reg_pos</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s1">&#39;Invalid region text provided.&#39;</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">error_message_and_exit</span><span class="p">(</span>
                <span class="s1">&#39;Invalid [--include-region] format.&#39;</span><span class="p">)</span>

        <span class="n">parsed_regs</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reg_pos</span><span class="p">)</span>

    <span class="n">parsed_regs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">parsed_regs</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="n">include_whole_chrms</span><span class="p">:</span>
        <span class="n">parsed_regs</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="n">parsed_regs</span>

<span class="k">def</span> <span class="nf">parse_locs_file</span><span class="p">(</span><span class="n">locs_fn</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parse BED files containing genomic locations (assumes single base locations, so end coordinate is ignored).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">raw_locs</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">locs_fn</span><span class="p">)</span> <span class="k">as</span> <span class="n">locs_fp</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">locs_fp</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">chrm</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">strand</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[:</span><span class="mi">6</span><span class="p">]</span>
                <span class="c1"># bed specs indicate 0-based start so don&#39;t shift here</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
                <span class="n">raw_locs</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">continue</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">((</span><span class="n">cs</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">cs_poss</span><span class="p">)))</span>
                <span class="k">for</span> <span class="n">cs</span><span class="p">,</span> <span class="n">cs_poss</span> <span class="ow">in</span> <span class="n">raw_locs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

<span class="k">def</span> <span class="nf">parse_obs_filter</span><span class="p">(</span><span class="n">obs_filter</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parse observations per base formatted filtering</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">obs_filter</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># parse obs_filter</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">obs_filter</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="n">pctl_nobs</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)))</span>
                      <span class="k">for</span> <span class="n">pctl_nobs</span> <span class="ow">in</span> <span class="n">obs_filter</span><span class="p">]</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s1">&#39;Invalid format for observation filter&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">pctl</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">pctl</span> <span class="o">&gt;</span> <span class="mi">100</span> <span class="k">for</span> <span class="n">pctl</span> <span class="ow">in</span> <span class="nb">map</span><span class="p">(</span><span class="n">itemgetter</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">obs_filter</span><span class="p">)):</span>
       <span class="n">error_message_and_exit</span><span class="p">(</span><span class="s1">&#39;Invalid percentile value.&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">obs_filter</span>

<span class="k">def</span> <span class="nf">get_seq_kmers</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">kmer_width</span><span class="p">,</span> <span class="n">rev_strand</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Compute expected signal levels for a sequence from a reference model</span>

<span class="sd">    Args:</span>
<span class="sd">        seq (str): genomic seqeunce to be converted to expected signal levels</span>
<span class="sd">        kmer_width (int): k-mer width</span>
<span class="sd">        rev_strand (bool): provided sequence is from reverse strand (so flip return order to genome forward direction)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_kmers</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="n">kmer_width</span><span class="p">]</span>
                 <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="n">kmer_width</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
    <span class="c1"># get stat lookups from seq on native strand then flip if rev_strand</span>
    <span class="k">if</span> <span class="n">rev_strand</span><span class="p">:</span>
        <span class="n">seq_kmers</span> <span class="o">=</span> <span class="n">seq_kmers</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">seq_kmers</span>

<div class="viewcode-block" id="TomboMotif"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboMotif">[docs]</a><span class="k">class</span> <span class="nc">TomboMotif</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Description of a sequence motif, including potentially modified position</span>

<span class="sd">    Attributes:</span>
<span class="sd">        raw_motif (str): input raw motif string</span>
<span class="sd">        motif_len (int): length of motif</span>
<span class="sd">        motif_pat (``re.compile``): python regular expression for this motif</span>
<span class="sd">        rev_comp_pat (``re.compile``): python regular expression for reverse complement of this motif</span>
<span class="sd">        is_palindrome (bool): is the motif palindromic (in the genomic not literary definition)</span>
<span class="sd">        mod_pos (int): modified base position within motif</span>
<span class="sd">        mod_base (str): modified base (should generally be A, C, G or T)</span>

<span class="sd">    .. automethod:: __init__</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">_parse_motif</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_motif</span><span class="p">,</span> <span class="n">rev_comp_motif</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="n">conv_motif</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">SINGLE_LETTER_CODE</span><span class="p">[</span><span class="n">letter</span><span class="p">]</span>
                             <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="n">raw_motif</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">rev_comp_motif</span><span class="p">:</span>
            <span class="c1"># reverse complement and then flip any group brackets</span>
            <span class="n">conv_motif</span> <span class="o">=</span> <span class="n">rev_comp</span><span class="p">(</span><span class="n">conv_motif</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">({</span>
                <span class="nb">ord</span><span class="p">(</span><span class="s1">&#39;[&#39;</span><span class="p">):</span><span class="s1">&#39;]&#39;</span><span class="p">,</span> <span class="nb">ord</span><span class="p">(</span><span class="s1">&#39;]&#39;</span><span class="p">):</span><span class="s1">&#39;[&#39;</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">conv_motif</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_compute_partial_patterns</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Compute patterns for partial matches that include the mod_pos</span>
<span class="sd">        at the start, end or within short sequences.</span>

<span class="sd">        Key into _partial_pats with:</span>
<span class="sd">            1) whether searching at start, end or within a short sequence</span>
<span class="sd">            2) length of the partial pattern</span>

<span class="sd">        Values are compiled partial pattern and mod_pos within pattern.</span>
<span class="sd">        Short patterns are lists.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;start&#39;</span><span class="p">:{},</span> <span class="s1">&#39;end&#39;</span><span class="p">:{},</span> <span class="s1">&#39;short&#39;</span><span class="p">:{}}</span>
        <span class="k">for</span> <span class="n">offset</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span> <span class="o">-</span> <span class="n">offset</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_parse_motif</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">raw_motif</span><span class="p">[</span><span class="n">offset</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:]),</span> <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span> <span class="o">-</span> <span class="n">offset</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">offset</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span> <span class="o">-</span> <span class="n">offset</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_parse_motif</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">raw_motif</span><span class="p">[:</span><span class="o">-</span><span class="p">(</span><span class="n">offset</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]),</span> <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">short_len</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span><span class="s1">&#39;short&#39;</span><span class="p">][</span><span class="n">short_len</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_parse_motif</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">raw_motif</span><span class="p">[</span><span class="n">offset</span><span class="p">:</span><span class="n">offset</span> <span class="o">+</span> <span class="n">short_len</span><span class="p">]),</span>
                 <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span> <span class="o">-</span> <span class="n">offset</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">offset</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span>
                        <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span> <span class="o">-</span> <span class="n">short_len</span><span class="p">),</span>
                        <span class="nb">min</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span> <span class="o">-</span> <span class="n">short_len</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span><span class="p">))]</span>
        <span class="k">return</span>

<div class="viewcode-block" id="TomboMotif.__init__"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboMotif.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_motif</span><span class="p">,</span> <span class="n">mod_pos</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parse string motif</span>

<span class="sd">        Args:</span>
<span class="sd">            raw_motif (str): sequence motif. supports IUPAC single letter codes (use T for RNA).</span>
<span class="sd">            mod_pos (int): 1-based position of modified base within the motif</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO convert mod_pos to 0-based coordinate</span>
        <span class="c1"># (1-based is much more error prone)</span>
        <span class="n">invalid_chars</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span>
            <span class="s1">&#39;[^&#39;</span> <span class="o">+</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">SINGLE_LETTER_CODE</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;]&#39;</span><span class="p">,</span> <span class="n">raw_motif</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">invalid_chars</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">error_message_and_exit</span><span class="p">(</span>
                <span class="s1">&#39;Invalid characters in motif: &#39;</span> <span class="o">+</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">invalid_chars</span><span class="p">))</span>

        <span class="c1"># basic motif parsing</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">raw_motif</span> <span class="o">=</span> <span class="n">raw_motif</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">raw_motif</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_pat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parse_motif</span><span class="p">(</span><span class="n">raw_motif</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rev_comp_pat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parse_motif</span><span class="p">(</span><span class="n">raw_motif</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">is_palindrome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_pat</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">rev_comp_pat</span>

        <span class="c1"># parse modified position from motif if provided</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span> <span class="o">=</span> <span class="n">mod_pos</span>
        <span class="k">if</span> <span class="n">mod_pos</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mod_base</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="mi">0</span> <span class="o">&lt;</span> <span class="n">mod_pos</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mod_base</span> <span class="o">=</span> <span class="n">raw_motif</span><span class="p">[</span><span class="n">mod_pos</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">INVALID_BASES</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mod_base</span><span class="p">):</span>
                <span class="n">warning_message</span><span class="p">(</span>
                    <span class="s1">&#39;Provided modified position is not a single base, which &#39;</span> <span class="o">+</span>
                    <span class="s1">&#39;is likely an error. Specified modified base is one of: &#39;</span> <span class="o">+</span>
                    <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">SINGLE_LETTER_CODE</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">mod_base</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_compute_partial_patterns</span><span class="p">()</span></div>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="s1">&#39;Raw Motif:</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">raw_motif</span><span class="p">,</span>
                          <span class="s1">&#39;Mod Position:</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span><span class="p">),</span>
                          <span class="s1">&#39;Motif Pattern:</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_pat</span><span class="p">),</span>
                          <span class="s1">&#39;Rev Comp Pattern:</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rev_comp_pat</span><span class="p">)))</span>

<div class="viewcode-block" id="TomboMotif.matches_seq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboMotif.matches_seq">[docs]</a>    <span class="k">def</span> <span class="nf">matches_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Does the motif match provided sequence (including mod_pos within seq)?</span>

<span class="sd">        Including partial matches at beginning and end that include mod_pos.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># check matches to start of sequence</span>
        <span class="k">for</span> <span class="n">start_len</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span><span class="p">)):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">start_pat</span><span class="p">,</span> <span class="n">start_mod_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span>
                    <span class="s1">&#39;start&#39;</span><span class="p">][</span><span class="n">start_len</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">start_pat</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">seq</span><span class="p">[:</span><span class="n">start_len</span><span class="p">]):</span>
                <span class="k">return</span> <span class="kc">True</span>

        <span class="c1"># check central sequence overlaps</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">short_pat</span><span class="p">,</span> <span class="n">mod_pos</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span><span class="s1">&#39;short&#39;</span><span class="p">][</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)]:</span>
                <span class="k">if</span> <span class="n">short_pat</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
                    <span class="k">return</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_pat</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
                <span class="k">return</span> <span class="kc">True</span>

        <span class="c1"># check end of seq matches</span>
        <span class="k">for</span> <span class="n">end_len</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span><span class="p">)):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">end_pat</span><span class="p">,</span> <span class="n">end_mod_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">][</span><span class="n">end_len</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">end_pat</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="o">-</span><span class="n">end_len</span><span class="p">:]):</span>
                <span class="k">return</span> <span class="kc">True</span>

        <span class="k">return</span> <span class="kc">False</span></div>

<div class="viewcode-block" id="TomboMotif.find_mod_poss"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboMotif.find_mod_poss">[docs]</a>    <span class="k">def</span> <span class="nf">find_mod_poss</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Find all mod-base positions within the sequence.</span>

<span class="sd">        Including partial matches at beginning and end that include mod_pos.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">seq_mod_poss</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="c1"># check matches to start of sequence</span>
        <span class="k">for</span> <span class="n">start_len</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span><span class="p">)):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">start_pat</span><span class="p">,</span> <span class="n">start_mod_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span>
                    <span class="s1">&#39;start&#39;</span><span class="p">][</span><span class="n">start_len</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">start_pat</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">seq</span><span class="p">[:</span><span class="n">start_len</span><span class="p">]):</span>
                <span class="n">seq_mod_poss</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">start_mod_pos</span><span class="p">)</span>

        <span class="c1"># check central sequence overlaps</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">short_pat</span><span class="p">,</span> <span class="n">short_mod_pos</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span>
                    <span class="s1">&#39;short&#39;</span><span class="p">][</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)]:</span>
                <span class="k">if</span> <span class="n">short_pat</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
                    <span class="n">seq_mod_poss</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">short_mod_pos</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">motif_match</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_pat</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
                <span class="n">seq_mod_poss</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">motif_match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">mod_pos</span><span class="p">)</span>

        <span class="c1"># check end of seq matches</span>
        <span class="k">for</span> <span class="n">end_len</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_len</span><span class="p">)):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">end_pat</span><span class="p">,</span> <span class="n">end_mod_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_partial_pats</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">][</span><span class="n">end_len</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">end_pat</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="o">-</span><span class="n">end_len</span><span class="p">:]):</span>
                <span class="n">seq_mod_poss</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="n">end_len</span> <span class="o">+</span> <span class="n">end_mod_pos</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">seq_mod_poss</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="parse_motif_descs"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.parse_motif_descs">[docs]</a><span class="k">def</span> <span class="nf">parse_motif_descs</span><span class="p">(</span><span class="n">stat_motif_descs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parse string motif descriptions as defined by ``tombo plot roc --motif-descriptions``</span>

<span class="sd">    Args:</span>
<span class="sd">        stat_motif_descs (str): string motifs description (see ``tombo plot roc --motif-descriptions``)</span>

<span class="sd">    Returns:</span>
<span class="sd">        list of tuples with :class:`tombo.tombo_helper.TomboMotif` and motif/modification names</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parsed_motif_descs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">motif_desc</span> <span class="ow">in</span> <span class="n">stat_motif_descs</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;::&#39;</span><span class="p">):</span>
            <span class="n">raw_motif</span><span class="p">,</span> <span class="n">mod_pos</span><span class="p">,</span> <span class="n">mod_name</span> <span class="o">=</span> <span class="n">motif_desc</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)</span>
            <span class="n">motif</span> <span class="o">=</span> <span class="n">TomboMotif</span><span class="p">(</span><span class="n">raw_motif</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">mod_pos</span><span class="p">))</span>
            <span class="n">parsed_motif_descs</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">motif</span><span class="p">,</span> <span class="n">mod_name</span><span class="p">))</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="n">error_message_and_exit</span><span class="p">(</span>
            <span class="s1">&#39;Invalid motif decriptions format. Format descriptions as: &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;&quot;motif:mod_pos:name[::motif2:mod_pos2:name2...]&quot;.&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parsed_motif_descs</span></div>


<span class="c1">###########################</span>
<span class="c1">###### FASTA Parsing ######</span>
<span class="c1">###########################</span>

<span class="k">def</span> <span class="nf">get_rec_names</span><span class="p">(</span><span class="n">fasta_fn</span><span class="p">):</span>
    <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">fasta_fn</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_fp</span><span class="p">:</span>
        <span class="n">all_rec_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fasta_fp</span>
                       <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)]</span>

    <span class="k">return</span> <span class="n">all_rec_ids</span>

<div class="viewcode-block" id="Fasta"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.Fasta">[docs]</a><span class="k">class</span> <span class="nc">Fasta</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Fasta file sequence format wrapper class. Will load faidx via ``pyfaidx`` package if installed, else the fasta will be loaded into memory for sequence extraction.</span>

<span class="sd">    .. automethod:: __init__</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">_load_in_mem</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">genome_index</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">curr_id</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">curr_seq</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fasta_fn</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_fp</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fasta_fp</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">curr_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span>
                        <span class="n">curr_seq</span> <span class="ow">is</span> <span class="ow">not</span> <span class="p">[]):</span>
                        <span class="n">genome_index</span><span class="p">[</span><span class="n">curr_id</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">curr_seq</span><span class="p">)</span>
                    <span class="n">curr_seq</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">curr_id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">curr_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
            <span class="c1"># add last record</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">curr_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span>
                <span class="n">curr_seq</span> <span class="ow">is</span> <span class="ow">not</span> <span class="p">[]):</span>
                <span class="n">genome_index</span><span class="p">[</span><span class="n">curr_id</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">curr_seq</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">genome_index</span>

    <span class="k">def</span> <span class="nf">_index_contains_uridines</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_chrms</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_bases</span><span class="o">=</span><span class="mi">1000</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span><span class="p">:</span>
            <span class="c1"># check first N bases of the first M chrms for U characters</span>
            <span class="k">for</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="n">islice</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">n_chrms</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s1">&#39;U&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span>
                        <span class="n">chrm</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_bases</span><span class="p">,</span> <span class="n">error_end</span><span class="o">=</span><span class="kc">False</span><span class="p">)):</span>
                    <span class="k">return</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="n">islice</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">n_chrms</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s1">&#39;U&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span>
                        <span class="n">chrm</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_bases</span><span class="p">,</span> <span class="n">error_end</span><span class="o">=</span><span class="kc">False</span><span class="p">)):</span>
                    <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="kc">False</span>

<div class="viewcode-block" id="Fasta.__init__"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.Fasta.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_fn</span><span class="p">,</span> <span class="n">dry_run</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">force_in_mem</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">assume_dna_base</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Load a fasta</span>

<span class="sd">        Args:</span>
<span class="sd">            fasta_fn (str): path to fasta file</span>
<span class="sd">            dry_run (bool): when pyfaidx is not installed, don&#39;t actually read sequence into memory.</span>
<span class="sd">            force_in_mem (bool): force genome to be loaded into memory even if pyfaidx is installed allowing on-disk access</span>
<span class="sd">            assume_dna_bases (bool): skip check for DNA or RNA bases (default: False)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fasta_fn</span> <span class="o">=</span> <span class="n">resolve_path</span><span class="p">(</span><span class="n">fasta_fn</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">has_rna_bases</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">force_in_mem</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ImportError</span>
            <span class="kn">import</span> <span class="nn">pyfaidx</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Faidx</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fasta_fn</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">UnicodeDecodeError</span><span class="p">:</span>
                <span class="n">error_message_and_exit</span><span class="p">(</span>
                    <span class="s1">&#39;FASTA file does not appear to be formatted correctly.&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">dry_run</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_in_mem</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">dry_run</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">has_rna_bases</span> <span class="o">=</span> <span class="p">(</span><span class="n">assume_dna_base</span> <span class="ow">or</span>
                                  <span class="bp">self</span><span class="o">.</span><span class="n">_index_contains_uridines</span><span class="p">())</span></div>

<div class="viewcode-block" id="Fasta.get_seq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.Fasta.get_seq">[docs]</a>    <span class="k">def</span> <span class="nf">get_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">error_end</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Extract sequence from a specific genomic region. Note if provided, start and end must both be provided or they will be ignored.</span>

<span class="sd">        Args:</span>

<span class="sd">            chrm (str): chromosome name</span>
<span class="sd">            start (int): 0-based start position</span>
<span class="sd">            end (int): 0-based open-interval end position</span>
<span class="sd">            error_end (bool): raise an error when the region requested extends beyond the chromosome (default: True)</span>

<span class="sd">        Returns:</span>
<span class="sd">            Genomic sequence requested. Sequence is converted to RNA sequence if applicable.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">start</span> <span class="ow">or</span> <span class="n">end</span><span class="p">):</span>
                <span class="n">r_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span>
                    <span class="n">chrm</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span><span class="o">.</span><span class="n">rlen</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="k">elif</span> <span class="p">(</span><span class="n">start</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span><span class="o">.</span><span class="n">rlen</span> <span class="ow">or</span> <span class="p">(</span>
                    <span class="n">error_end</span> <span class="ow">and</span> <span class="p">(</span>
                        <span class="n">end</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">]</span><span class="o">.</span><span class="n">rlen</span><span class="p">))):</span>
                <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span>
                    <span class="s1">&#39;Encountered invalid genome sequence request.&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">r_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="p">(</span>
                    <span class="n">start</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">])))</span> <span class="ow">or</span> <span class="p">(</span>
                        <span class="n">error_end</span> <span class="ow">and</span> <span class="n">end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span>
                        <span class="p">(</span><span class="n">end</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span>
                         <span class="n">end</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">]))):</span>
                <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span>
                    <span class="s1">&#39;Encountered invalid genome sequence request.&#39;</span><span class="p">)</span>
            <span class="n">r_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">chrm</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_rna_bases</span><span class="p">:</span>
            <span class="n">r_seq</span> <span class="o">=</span> <span class="n">rev_transcribe</span><span class="p">(</span><span class="n">r_seq</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">r_seq</span></div>

<div class="viewcode-block" id="Fasta.iter_chrms"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.Fasta.iter_chrms">[docs]</a>    <span class="k">def</span> <span class="nf">iter_chrms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Iterate over chromosome names</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">unicode</span><span class="p">(</span><span class="n">chrm</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">chrm</span></div>

    <span class="k">def</span> <span class="nf">__contains__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_pyfaidx</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">index</span>
        <span class="k">return</span> <span class="n">chrm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span></div>


<span class="c1">#############################################</span>
<span class="c1">###### Automatic Sample-Type Detection ######</span>
<span class="c1">#############################################</span>

<span class="k">def</span> <span class="nf">is_read_rna</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Determine if a read is RNA or DNA</span>

<span class="sd">    Args:</span>
<span class="sd">        fast5_data (`h5py.File`)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Boolean, incdicating whether the read appears to be RNA or DNA</span>

<span class="sd">    Note:</span>
<span class="sd">        This function uses the read meta-data and so non-standard processing pipelines may not prodcue expected values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># check both experiment type and kit slots for &quot;rna&quot;</span>
    <span class="n">exp_type</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;UniqueGlobalKey/context_tags&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="s1">&#39;experiment_type&#39;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">exp_type</span> <span class="o">=</span> <span class="n">exp_type</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
        <span class="c1"># remove the word internal since it contains rna.</span>
        <span class="n">exp_type</span> <span class="o">=</span> <span class="n">exp_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;internal&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
        <span class="k">pass</span>

    <span class="n">exp_kit</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;UniqueGlobalKey/context_tags&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="s1">&#39;experiment_kit&#39;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">exp_kit</span> <span class="o">=</span> <span class="n">exp_kit</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
        <span class="c1"># remove the word internal since it contains rna.</span>
        <span class="n">exp_kit</span> <span class="o">=</span> <span class="n">exp_kit</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;internal&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
        <span class="k">pass</span>

    <span class="k">if</span> <span class="n">exp_type</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">exp_kit</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="p">(</span><span class="n">exp_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span> <span class="n">exp_type</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span>
        <span class="p">(</span><span class="n">exp_kit</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span> <span class="n">exp_kit</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">))</span>

<span class="k">def</span> <span class="nf">is_sample_rna</span><span class="p">(</span><span class="n">reads_index</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fast5_fns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n_reads</span><span class="o">=</span><span class="mi">50</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Determine if a set of reads are RNA or DNA from a small sample. Must provide either reads_index or fast5_fns.</span>

<span class="sd">    Args:</span>
<span class="sd">        reads_index (:class:`tombo.tombo_helper.TomboReads`)</span>
<span class="sd">        fast5_fns (list): list of fast5 read filename</span>
<span class="sd">        n_reads (int): number of reads to check (default: 50)</span>

<span class="sd">    Returns:</span>
<span class="sd">        False if any read does not appear to be an RNA read (see :class:`tombo.tombo_helper.is_read_rna`) else return True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">proc_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">reads_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">reads_index</span><span class="o">.</span><span class="n">iter_reads</span><span class="p">():</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">r_data</span><span class="o">.</span><span class="n">rna</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
            <span class="n">proc_reads</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">proc_reads</span> <span class="o">&gt;=</span> <span class="n">n_reads</span><span class="p">:</span>
                <span class="k">break</span>
    <span class="k">elif</span> <span class="n">fast5_fns</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">fast5_fn</span> <span class="ow">in</span> <span class="n">fast5_fns</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">is_read_rna</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">):</span>
                        <span class="k">return</span> <span class="kc">False</span>
                <span class="n">proc_reads</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">proc_reads</span> <span class="o">&gt;=</span> <span class="n">n_reads</span><span class="p">:</span>
                <span class="k">break</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span>
            <span class="s1">&#39;Must provide either reads_index or fast5_fns to &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;determine is sample is RNA.&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="kc">True</span>

<div class="viewcode-block" id="get_seq_sample_type"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_seq_sample_type">[docs]</a><span class="k">def</span> <span class="nf">get_seq_sample_type</span><span class="p">(</span><span class="n">fast5_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reads_index</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fast5_fns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                        <span class="n">num_reads</span><span class="o">=</span><span class="mi">50</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get the sequencing sample type from a single read or set of reads</span>

<span class="sd">    Args:</span>
<span class="sd">        fast5_data (`h5py.File`): open read h5py File object</span>
<span class="sd">        reads_index (:class:`tombo.tombo_helper.TomboReads`)</span>
<span class="sd">        fast5_fns (list): FAST5 read filenames</span>
<span class="sd">        num_reads (int): sample of reads to check for sample type</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">fast5_data</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">reads_index</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">fast5_fns</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s1">&#39;Must provide either fast5_data, reads_index, or &#39;</span> <span class="o">+</span>
                         <span class="s1">&#39;fast5_fns to determine sequencing sample type.&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">fast5_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">seqSampleType</span><span class="p">(</span><span class="n">RNA_SAMP_TYPE</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span> <span class="k">if</span> <span class="n">is_sample_rna</span><span class="p">(</span>
            <span class="n">reads_index</span><span class="p">,</span> <span class="n">fast5_fns</span><span class="p">,</span> <span class="n">num_reads</span><span class="p">)</span> <span class="k">else</span> <span class="n">seqSampleType</span><span class="p">(</span>
                <span class="n">DNA_SAMP_TYPE</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seqSampleType</span><span class="p">(</span><span class="n">RNA_SAMP_TYPE</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span> <span class="k">if</span> <span class="n">is_read_rna</span><span class="p">(</span>
        <span class="n">fast5_data</span><span class="p">)</span> <span class="k">else</span> <span class="n">seqSampleType</span><span class="p">(</span><span class="n">DNA_SAMP_TYPE</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span></div>


<span class="c1">############################</span>
<span class="c1">###### Lock Functions ######</span>
<span class="c1">############################</span>

<span class="k">def</span> <span class="nf">get_lock_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get filename for the lock file to indicate that this directory is currently being processed. This file should be saved to be deleted later.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># if directory comes with trailing slash, remove for processing</span>
    <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">):</span>
        <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="n">fast5s_dir</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">split_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">split_dir</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;.&quot;</span> <span class="o">+</span> <span class="n">split_dir</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.tombo.lock&#39;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">_is_lock_file</span><span class="p">(</span><span class="n">lock_fn</span><span class="p">):</span>
    <span class="n">lock_bn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">lock_fn</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lock_bn</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span>
        <span class="ow">not</span> <span class="n">lock_bn</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span> <span class="ow">or</span>
        <span class="ow">not</span> <span class="n">lock_bn</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.tombo.lock&#39;</span><span class="p">)):</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">return</span> <span class="kc">True</span>


<span class="c1">#####################################</span>
<span class="c1">###### FAST5 Parsing Functions ######</span>
<span class="c1">#####################################</span>

<span class="k">def</span> <span class="nf">reads_contain_basecalls</span><span class="p">(</span><span class="n">fast5_fns</span><span class="p">,</span> <span class="n">bc_grp</span><span class="p">,</span> <span class="n">num_reads</span><span class="p">):</span>
    <span class="n">test_fns</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span>
        <span class="n">fast5_fns</span><span class="p">,</span> <span class="n">num_reads</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fast5_fns</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">num_reads</span> <span class="k">else</span> <span class="n">fast5_fns</span>
    <span class="k">for</span> <span class="n">fast5_fn</span> <span class="ow">in</span> <span class="n">test_fns</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
                <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">bc_grp</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="c1"># if the basecall group is accessible for a single file return true</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="c1"># else if all tested reads did not contain the basecall group return False</span>
    <span class="k">return</span> <span class="kc">False</span>

<span class="k">def</span> <span class="nf">get_files_list</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get all fast5 files recursively below this directory</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_fast5s</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># walk through directory structure searching for fast5 files</span>
    <span class="k">for</span> <span class="n">root</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fns</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">walk</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">fns</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">fn</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fast5&#39;</span><span class="p">):</span> <span class="k">continue</span>
            <span class="n">all_fast5s</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">fn</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">all_fast5s</span>

<span class="k">def</span> <span class="nf">clear_tombo_locks</span><span class="p">(</span><span class="n">lock_fns</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Clear all lock files</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">lock_fn</span> <span class="ow">in</span> <span class="n">lock_fns</span><span class="p">:</span>
        <span class="c1"># safegaurd against incorrect file passed to this function</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">_is_lock_file</span><span class="p">(</span><span class="n">lock_fn</span><span class="p">):</span>
            <span class="k">continue</span>
        <span class="c1"># try to remove lock files so that is some have been corrupted or</span>
        <span class="c1"># otherwise cause an error at least all those accessible will be cleared</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">lock_fn</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">pass</span>

    <span class="k">return</span>

<span class="k">def</span> <span class="nf">get_files_list_and_lock_dirs</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">ignore_locks</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get all fast5 files recursively below this directory and add a Tombo lock file to indicate that this directory is currently being re-squiggled</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ignore_locks_mess</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s1">&#39;This set of reads is currently being processed by another &#39;</span> <span class="o">+</span>
        <span class="s1">&#39;resquiggle command. Multiple resquiggle commands cannot be &#39;</span> <span class="o">+</span>
        <span class="s1">&#39;run concurrently on a set of reads to avoid corrupting &#39;</span> <span class="o">+</span>
        <span class="s1">&#39;read files. If you are sure this set of reads is not being &#39;</span> <span class="o">+</span>
        <span class="s1">&#39;processed by another command (usually caused by previous &#39;</span> <span class="o">+</span>
        <span class="s1">&#39;unexpected exit) set the --ignore-read-locks flag.&#39;</span><span class="p">)</span>
    <span class="n">all_fast5s</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">lock_fns</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># walk through directory structure searching for fast5 files</span>
        <span class="k">for</span> <span class="n">root</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fns</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">walk</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">):</span>
            <span class="n">lock_fn</span> <span class="o">=</span> <span class="n">get_lock_fn</span><span class="p">(</span><span class="n">root</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">ignore_locks</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">lock_fn</span><span class="p">):</span>
                <span class="n">clear_tombo_locks</span><span class="p">(</span><span class="n">lock_fns</span><span class="p">)</span>
                <span class="n">error_message_and_exit</span><span class="p">(</span><span class="n">ignore_locks_mess</span><span class="p">)</span>
            <span class="n">lock_fns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lock_fn</span><span class="p">)</span>
            <span class="c1"># create empty file indicating this directory is locked</span>
            <span class="nb">open</span><span class="p">(</span><span class="n">lock_fn</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

            <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">fns</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">fn</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fast5&#39;</span><span class="p">):</span> <span class="k">continue</span>
                <span class="n">all_fast5s</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">fn</span><span class="p">))</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="n">clear_tombo_locks</span><span class="p">(</span><span class="n">lock_fns</span><span class="p">)</span>
        <span class="n">error_message_and_exit</span><span class="p">(</span>
            <span class="s1">&#39;Unexpected error during file enumeration. Check that you have &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;write permission within the specified [fast5_basedir].&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">all_fast5s</span><span class="p">,</span> <span class="n">lock_fns</span>

<div class="viewcode-block" id="get_raw_read_slot"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_raw_read_slot">[docs]</a><span class="k">def</span> <span class="nf">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get the raw read slot from this FAST5 read file</span>

<span class="sd">    Args:</span>
<span class="sd">        fast5_data (`h5py.File`): open FAST5 read file object</span>

<span class="sd">    Example::</span>

<span class="sd">        all_aw_signal = get_raw_read_slot(fast5_data)[&#39;Signal&#39;][:]</span>

<span class="sd">    Returns:</span>
<span class="sd">        The HDF5 group slot containing the raw signal data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">raw_read_slot</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Raw/Reads&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s1">&#39;Raw data is not found in /Raw/Reads/Read_[read#]&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">raw_read_slot</span></div>

<div class="viewcode-block" id="TomboReads"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads">[docs]</a><span class="k">class</span> <span class="nc">TomboReads</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;A set of reads with associated meta-data from re-squiggle processing</span>

<span class="sd">    .. automethod:: __init__</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">_get_index_fn</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get the filename for the requested directory and corrected group</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># if directory comes with trailing slash, remove for processing</span>
        <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">):</span>
            <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="n">fast5s_dir</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">split_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">split_dir</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;.&quot;</span> <span class="o">+</span> <span class="n">split_dir</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span>
                            <span class="s2">&quot;.&quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span> <span class="o">+</span> <span class="s1">&#39;.tombo.index&#39;</span><span class="p">)</span>


    <span class="c1"># index building and writing class functions</span>
    <span class="k">def</span> <span class="nf">_prep_for_writing</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fast5s_dirs</span><span class="p">):</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">fast5s_dirs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="p">(</span>
            <span class="s1">&#39;Must provide only a single FAST5 base directory when &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;openning for writing.&#39;</span><span class="p">)</span>
        <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="n">fast5s_dirs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">fast5s_dir</span> <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
                      <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
        <span class="n">index_fn</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fast5s_dirs</span><span class="p">[</span><span class="n">fast5s_dir</span><span class="p">]</span> <span class="o">=</span> <span class="n">index_fn</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">index_fn</span><span class="p">):</span> <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">index_fn</span><span class="p">)</span>

        <span class="c1"># open default dict to fill with readData lists by (chrm, strand)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

        <span class="k">return</span>

<div class="viewcode-block" id="TomboReads.add_read_data"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.add_read_data">[docs]</a>    <span class="k">def</span> <span class="nf">add_read_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">read_data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add read data to the index</span>

<span class="sd">        Args:</span>
<span class="sd">            chrm (str): chromosome name</span>
<span class="sd">            strand (str): strand (&#39;+&#39; or &#39;-&#39;)</span>
<span class="sd">            read_data (:class:`tombo.tombo_helper.readData`): read information</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read_data</span><span class="p">)</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="TomboReads.replace_index"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.replace_index">[docs]</a>    <span class="k">def</span> <span class="nf">replace_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_reads_index</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Replace current reads index</span>

<span class="sd">        Args:</span>
<span class="sd">            new_reads_index (dict): dictionary with (chrm, strand) pointing to lists of :class:`tombo.tombo_helper.readData` objects</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_reads_index</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s1">&#39;Cannot replace with an empty index.&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span> <span class="o">=</span> <span class="n">new_reads_index</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="TomboReads.write_index_file"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.write_index_file">[docs]</a>    <span class="k">def</span> <span class="nf">write_index_file</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Write index file</span>

<span class="sd">        Note:</span>

<span class="sd">            Must be an index from only a single ``fast5s_dir``</span>

<span class="sd">            Index filename will be: `.[fast5s_dir].[corr_grp].tombo.index`</span>
<span class="sd">            Note that this is a hidden file (starts with a &quot;.&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">status_message</span><span class="p">(</span><span class="s1">&#39;Saving Tombo reads index to file.&#39;</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fast5s_dirs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="p">(</span>
            <span class="s1">&#39;Cannot write index for TomboReads created from more than &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;one base directory.&#39;</span><span class="p">)</span>
        <span class="n">basedir</span><span class="p">,</span> <span class="n">index_fn</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fast5s_dirs</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="kn">import</span> <span class="nn">cPickle</span> <span class="k">as</span> <span class="nn">pickle</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="kn">import</span> <span class="nn">pickle</span>

        <span class="n">index_data</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chrm_strand</span><span class="p">,</span> <span class="n">cs_reads</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">rd</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">:</span>
                <span class="c1"># clip the basedir off the FAST5 filename in case later</span>
                <span class="c1"># functions are called from another relative path and</span>
                <span class="c1"># split corr_grp and bc_subgrp for easier filtering</span>
                <span class="n">index_data</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span>
                    <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">basedir</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">rd</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">rd</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">rd</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                    <span class="n">rd</span><span class="o">.</span><span class="n">read_start_rel_to_raw</span><span class="p">,</span> <span class="n">rd</span><span class="o">.</span><span class="n">corr_group</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
                    <span class="n">rd</span><span class="o">.</span><span class="n">corr_group</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">rd</span><span class="o">.</span><span class="n">filtered</span><span class="p">,</span> <span class="n">rd</span><span class="o">.</span><span class="n">rna</span><span class="p">,</span>
                    <span class="n">rd</span><span class="o">.</span><span class="n">sig_match_score</span><span class="p">,</span> <span class="n">rd</span><span class="o">.</span><span class="n">mean_q_score</span><span class="p">,</span> <span class="n">rd</span><span class="o">.</span><span class="n">read_id</span><span class="p">))</span>

        <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">index_fn</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">index_fp</span><span class="p">:</span>
            <span class="c1"># note protocol 2 for py2/3 compatibility</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">index_data</span><span class="p">),</span> <span class="n">index_fp</span><span class="p">,</span> <span class="n">protocol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

        <span class="k">return</span></div>


    <span class="c1"># Index parsing class functions</span>
    <span class="k">def</span> <span class="nf">_parse_fast5s_wo_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">wo_index_fast5s_dirs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parse re-squiggled reads data from a list of fast5 directories</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">def</span> <span class="nf">get_read_data</span><span class="p">(</span><span class="n">read_fn</span><span class="p">,</span> <span class="n">fast5_data</span><span class="p">,</span> <span class="n">bc_subgrp</span><span class="p">):</span>
            <span class="n">read_id</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;read_id&#39;</span><span class="p">)</span>
            <span class="n">corr_data</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span>
                <span class="s1">&#39;/&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span><span class="p">,</span> <span class="n">bc_subgrp</span><span class="p">))]</span>
            <span class="n">rna</span> <span class="o">=</span> <span class="n">corr_data</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;rna&#39;</span><span class="p">)</span>
            <span class="n">rna</span> <span class="o">=</span> <span class="kc">False</span> <span class="k">if</span> <span class="n">rna</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">rna</span>

            <span class="n">align_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">corr_data</span><span class="p">[</span><span class="s1">&#39;Alignment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
            <span class="n">read_start_rel_to_raw</span> <span class="o">=</span> <span class="n">corr_data</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                <span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">)</span>
            <span class="n">chrm</span> <span class="o">=</span> <span class="n">align_data</span><span class="p">[</span><span class="s1">&#39;mapped_chrom&#39;</span><span class="p">]</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="n">align_data</span><span class="p">[</span><span class="s1">&#39;mapped_strand&#39;</span><span class="p">]</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">chrm</span> <span class="o">=</span> <span class="n">chrm</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="n">strand</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">pass</span>

            <span class="k">return</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">readData</span><span class="p">(</span>
                <span class="n">align_data</span><span class="p">[</span><span class="s1">&#39;mapped_start&#39;</span><span class="p">],</span> <span class="n">align_data</span><span class="p">[</span><span class="s1">&#39;mapped_end&#39;</span><span class="p">],</span>
                <span class="kc">False</span><span class="p">,</span> <span class="n">read_start_rel_to_raw</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">read_fn</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">bc_subgrp</span><span class="p">,</span> <span class="n">rna</span><span class="p">,</span> <span class="n">read_id</span><span class="o">=</span><span class="n">read_id</span><span class="p">)</span>


        <span class="n">files</span> <span class="o">=</span> <span class="p">[</span><span class="n">fn</span> <span class="k">for</span> <span class="n">fast5s_dir</span> <span class="ow">in</span> <span class="n">wo_index_fast5s_dirs</span>
                 <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">get_files_list</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)]</span>
        <span class="n">dir_reads_index</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">read_fn</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">read_fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
                    <span class="n">i_bc_subgrps</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">bc_subgrps</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">bc_subgrps</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">bc_subgrp</span> <span class="ow">in</span> <span class="n">i_bc_subgrps</span><span class="p">:</span>
                        <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">r_data</span> <span class="o">=</span> <span class="n">get_read_data</span><span class="p">(</span>
                            <span class="n">read_fn</span><span class="p">,</span> <span class="n">fast5_data</span><span class="p">,</span> <span class="n">bc_subgrp</span><span class="p">)</span>
                        <span class="n">dir_reads_index</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r_data</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="c1"># ignore errors and process all reads that don&#39;t error</span>
                <span class="k">continue</span>

        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">dir_reads_index</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_load_index_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="kn">import</span> <span class="nn">cPickle</span> <span class="k">as</span> <span class="nn">pickle</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="kn">import</span> <span class="nn">pickle</span>
        <span class="k">with</span> <span class="n">io</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fast5s_dirs</span><span class="p">[</span><span class="n">fast5s_dir</span><span class="p">],</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">index_fp</span><span class="p">:</span>
            <span class="n">raw_index_data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">index_fp</span><span class="p">)</span>

        <span class="c1"># determine the index type used. Index information was added around</span>
        <span class="c1"># version 1.3 making the index data 2 elements longer.</span>
        <span class="c1"># so check which one is used here.</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">num_index_vals</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">raw_index_data</span><span class="o">.</span><span class="n">values</span><span class="p">()))[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">except</span> <span class="ne">StopIteration</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s1">&#39;Tombo index file appears to be empty&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_index_vals</span> <span class="o">==</span> <span class="mi">8</span><span class="p">:</span>
            <span class="k">def</span> <span class="nf">convert_r_data</span><span class="p">(</span><span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span>
                               <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">readData</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span>
                                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">from_base_fn</span><span class="p">),</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">rna</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">num_index_vals</span> <span class="o">==</span> <span class="mi">10</span><span class="p">:</span>
            <span class="k">def</span> <span class="nf">convert_r_data</span><span class="p">(</span>
                    <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">,</span>
                    <span class="n">sig_match_score</span><span class="p">,</span> <span class="n">mean_q_score</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">readData</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span>
                                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">from_base_fn</span><span class="p">),</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">rna</span><span class="p">,</span>
                                <span class="n">sig_match_score</span><span class="p">,</span> <span class="n">mean_q_score</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">num_index_vals</span> <span class="o">==</span> <span class="mi">11</span><span class="p">:</span>
            <span class="k">def</span> <span class="nf">convert_r_data</span><span class="p">(</span>
                    <span class="n">from_base_fn</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">c_grp</span><span class="p">,</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rna</span><span class="p">,</span>
                    <span class="n">sig_match_score</span><span class="p">,</span> <span class="n">mean_q_score</span><span class="p">,</span> <span class="n">read_id</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">readData</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">filtered</span><span class="p">,</span> <span class="n">rsrtr</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span>
                                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">,</span> <span class="n">from_base_fn</span><span class="p">),</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">s_grp</span><span class="p">,</span> <span class="n">rna</span><span class="p">,</span>
                                <span class="n">sig_match_score</span><span class="p">,</span> <span class="n">mean_q_score</span><span class="p">,</span> <span class="n">read_id</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s1">&#39;Invalid Tombo index file.&#39;</span><span class="p">)</span>

        <span class="n">dir_reads_index</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_raw_data</span> <span class="ow">in</span> <span class="n">raw_index_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">cs_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">convert_r_data</span><span class="p">(</span><span class="o">*</span><span class="n">r_data</span><span class="p">)</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_raw_data</span><span class="p">]</span>
            <span class="c1"># don&#39;t add chrm/strand if all reads are filtered</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs_data</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">dir_reads_index</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span> <span class="o">=</span> <span class="n">cs_data</span>

        <span class="k">return</span> <span class="n">dir_reads_index</span>

    <span class="k">def</span> <span class="nf">_parse_fast5s_w_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fast5s_dir</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Use index file to parse information about a set of reads</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">curr_dir_reads_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_index_data</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">UnicodeDecodeError</span><span class="p">:</span>
            <span class="n">warning_message</span><span class="p">(</span>
                <span class="s1">&#39;Invalid Tombo index file.</span><span class="se">\n\t\t</span><span class="s1">This occurs most often &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;when the re-squiggle command was completed using a Tombo &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;build against a different python version (2 or 3).&#39;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">TomboError</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_filtered</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">bc_subgrps</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">curr_dir_reads_index</span>

        <span class="n">filt_dir_reads_index</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_raw_data</span> <span class="ow">in</span> <span class="n">curr_dir_reads_index</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">cs_data</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">rd</span> <span class="k">for</span> <span class="n">rd</span> <span class="ow">in</span> <span class="n">cs_raw_data</span>
                <span class="k">if</span> <span class="n">rd</span><span class="o">.</span><span class="n">corr_group</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span> <span class="ow">and</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bc_subgrps</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span>
                 <span class="n">rd</span><span class="o">.</span><span class="n">corr_group</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">bc_subgrps</span><span class="p">)</span> <span class="ow">and</span>
                <span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_filtered</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">rd</span><span class="o">.</span><span class="n">filtered</span><span class="p">)]</span>
            <span class="c1"># don&#39;t add chrm/strand if all reads are filtered</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs_data</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">filt_dir_reads_index</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span> <span class="o">=</span> <span class="n">cs_data</span>

        <span class="k">return</span> <span class="n">filt_dir_reads_index</span>

    <span class="k">def</span> <span class="nf">_merge_dir_indices</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">w_index_ri</span><span class="p">,</span> <span class="n">wo_index_ri</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Merge coverage from serveral parsed sets of data</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_raw_reads_index</span> <span class="o">=</span> <span class="n">w_index_ri</span> <span class="o">+</span> <span class="p">[</span><span class="n">wo_index_ri</span><span class="p">,]</span>
        <span class="n">reads_index</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chrm_strand</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">([</span><span class="n">cs</span> <span class="k">for</span> <span class="n">d_ri</span> <span class="ow">in</span> <span class="n">all_raw_reads_index</span>
                                <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="n">d_ri</span><span class="p">]):</span>
            <span class="k">for</span> <span class="n">dir_reads_index</span> <span class="ow">in</span> <span class="n">all_raw_reads_index</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">chrm_strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dir_reads_index</span><span class="p">:</span> <span class="k">continue</span>
                <span class="n">reads_index</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">dir_reads_index</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">])</span>

        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">reads_index</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_parse_fast5s</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fast5s_dirs</span><span class="p">):</span>
        <span class="n">wo_index_dirs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">w_index_covs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">warn_index</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="c1"># determine if index exists for each directory and load appropriately</span>
        <span class="k">for</span> <span class="n">fast5s_dir</span> <span class="ow">in</span> <span class="n">fast5s_dirs</span><span class="p">:</span>
            <span class="n">fast5s_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">fast5s_dir</span> <span class="k">if</span> <span class="n">fast5s_dir</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="k">else</span>
                          <span class="n">fast5s_dir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fast5s_dirs</span><span class="p">[</span><span class="n">fast5s_dir</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_index_fn</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fast5s_dirs</span><span class="p">[</span><span class="n">fast5s_dir</span><span class="p">]):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">w_index_covs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_parse_fast5s_w_index</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">))</span>
                <span class="k">except</span> <span class="n">TomboError</span><span class="p">:</span>
                    <span class="n">warning_message</span><span class="p">(</span>
                        <span class="s1">&#39;Failed to parse tombo index file for &#39;</span> <span class="o">+</span> <span class="n">fast5s_dir</span> <span class="o">+</span>
                        <span class="s1">&#39; directory. Creating temporary index from &#39;</span> <span class="o">+</span>
                        <span class="s1">&#39;FAST5 files.&#39;</span><span class="p">)</span>
                    <span class="n">wo_index_dirs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">warn_index</span><span class="p">:</span>
                    <span class="n">warning_message</span><span class="p">(</span>
                        <span class="s1">&#39;Tombo index file does not exist for one or more &#39;</span> <span class="o">+</span>
                        <span class="s1">&#39;directories.</span><span class="se">\n\t\t</span><span class="s1">If --skip-index was not set for &#39;</span> <span class="o">+</span>
                        <span class="s1">&#39;re-squiggle command, ensure that the specified &#39;</span> <span class="o">+</span>
                        <span class="s1">&#39;directory is the same as for the re-squiggle command.&#39;</span><span class="p">)</span>
                    <span class="n">warn_index</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">wo_index_dirs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fast5s_dir</span><span class="p">)</span>

        <span class="n">wo_index_cov</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parse_fast5s_wo_index</span><span class="p">(</span><span class="n">wo_index_dirs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_dir_indices</span><span class="p">(</span><span class="n">w_index_covs</span><span class="p">,</span> <span class="n">wo_index_cov</span><span class="p">)</span>

        <span class="k">return</span>

<div class="viewcode-block" id="TomboReads.__init__"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fast5s_basedirs</span><span class="p">,</span> <span class="n">corrected_group</span><span class="o">=</span><span class="s1">&#39;RawGenomeCorrected_000&#39;</span><span class="p">,</span>
                 <span class="n">basecall_subgroups</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">for_writing</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">remove_filtered</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                 <span class="n">sample_name</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parse data from a list of re-squiggle fast5 directories</span>

<span class="sd">        Args:</span>
<span class="sd">            fast5s_basedirs (list): fast5 base directories, which have been processed by ``tombo resquiggle``</span>
<span class="sd">            corrected_group (str): Analysis slot containing the re-squiggle information (optional; default: &#39;RawGenomeCorrected_000&#39;)</span>
<span class="sd">            basecall_subgroups (list): basecall subgroups (optional; default: Process all basecall subgroups)</span>
<span class="sd">            for_writing (bool): open TomboReads to write index (optional; default: Open for reading and parse re-squiggled reads)</span>
<span class="sd">            remove_filtered (bool): remove filtered reads as indicated by index file (optional; default: True)</span>
<span class="sd">            sample_name (str): for verbose output</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">VERBOSE</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">for_writing</span><span class="p">:</span>
            <span class="n">status_mess</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;Parsing Tombo index file(s).&#39;</span>
                           <span class="k">if</span> <span class="n">sample_name</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span>
                           <span class="s1">&#39;Parsing &#39;</span> <span class="o">+</span> <span class="n">sample_name</span> <span class="o">+</span> <span class="s1">&#39; Tombo index file(s).&#39;</span><span class="p">)</span>
            <span class="n">status_message</span><span class="p">(</span><span class="n">status_mess</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fast5s_basedirs</span><span class="p">,</span> <span class="nb">list</span><span class="p">),</span> <span class="p">(</span>
            <span class="s1">&#39;fast5s_basedirs must be a list.&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fast5s_dirs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">corr_grp</span> <span class="o">=</span> <span class="n">corrected_group</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bc_subgrps</span> <span class="o">=</span> <span class="n">basecall_subgroups</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">sample_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">remove_filtered</span> <span class="o">=</span> <span class="n">remove_filtered</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">for_writing</span> <span class="o">=</span> <span class="n">for_writing</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">for_writing</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_prep_for_writing</span><span class="p">(</span><span class="n">fast5s_basedirs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_parse_fast5s</span><span class="p">(</span><span class="n">fast5s_basedirs</span><span class="p">)</span>

        <span class="k">return</span></div>

    <span class="k">def</span> <span class="nf">_compute_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span> <span class="n">status_message</span><span class="p">(</span><span class="s1">&#39;Calculating read coverage.&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">chrm_strand</span><span class="p">,</span> <span class="n">cs_reads</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs_reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">cs_coverage</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">),</span>
                                   <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">:</span>
                <span class="n">cs_coverage</span><span class="p">[</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span><span class="n">r_data</span><span class="o">.</span><span class="n">end</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">cs_coverage</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">_add_coverages</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="p">):</span>
        <span class="n">merged_cov</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="c1"># if self or control reads don&#39;t have coverage, compute it</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_compute_coverage</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">ctrl_reads_index</span><span class="o">.</span><span class="n">coverage</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ctrl_reads_index</span><span class="o">.</span><span class="n">_compute_coverage</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">chrm_strand</span><span class="p">,</span> <span class="n">ctrl_cs_cov</span> <span class="ow">in</span> <span class="n">ctrl_reads_index</span><span class="o">.</span><span class="n">coverage</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">chrm_strand</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">:</span>
                <span class="n">self_cs_cov</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">]</span>
                <span class="c1"># copy longer array and add shorter array coverage</span>
                <span class="k">if</span> <span class="n">self_cs_cov</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">ctrl_cs_cov</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="n">merged_cs_cov</span> <span class="o">=</span> <span class="n">self_cs_cov</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                    <span class="n">merged_cs_cov</span><span class="p">[:</span><span class="n">ctrl_cs_cov</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+=</span> <span class="n">ctrl_cs_cov</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">merged_cs_cov</span> <span class="o">=</span> <span class="n">ctrl_cs_cov</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                    <span class="n">merged_cs_cov</span><span class="p">[:</span><span class="n">self_cs_cov</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+=</span> <span class="n">self_cs_cov</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">merged_cs_cov</span> <span class="o">=</span> <span class="n">ctrl_cs_cov</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">merged_cov</span><span class="p">[</span><span class="n">chrm_strand</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_cs_cov</span>

        <span class="k">return</span> <span class="n">merged_cov</span>

<div class="viewcode-block" id="TomboReads.iter_coverage_regions"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.iter_coverage_regions">[docs]</a>    <span class="k">def</span> <span class="nf">iter_coverage_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get genome coverage for a set of reads</span>

<span class="sd">        Args:</span>

<span class="sd">            ctrl_reads_index (:class:`tombo.tombo_helper.TomboReads`): a second set of tombo reads to add for coverage</span>

<span class="sd">        Returns:</span>
<span class="sd">            Yeilds (chrm, strand, cs_cov, cs_cov_starts) indicating read coverage levels</span>

<span class="sd">            cs_cov and cs_cov_starts are numpy arrays containing the coverage level and the 0-based genomic start positions of those intervals</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">VERBOSE</span><span class="p">:</span> <span class="n">status_message</span><span class="p">(</span><span class="s1">&#39;Calculating read coverage regions.&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_compute_coverage</span><span class="p">()</span>

        <span class="n">coverage</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="k">if</span> <span class="n">ctrl_reads_index</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_add_coverages</span><span class="p">(</span><span class="n">ctrl_reads_index</span><span class="p">))</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">cs_cov</span> <span class="ow">in</span> <span class="n">coverage</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">cs_cov_starts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,],</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">cs_cov</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                <span class="p">[</span><span class="n">cs_cov</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],]])</span>
            <span class="n">cs_cov</span> <span class="o">=</span> <span class="n">cs_cov</span><span class="p">[</span><span class="n">cs_cov_starts</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>
            <span class="k">yield</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">cs_cov</span><span class="p">,</span> <span class="n">cs_cov_starts</span>

        <span class="k">return</span></div>

<div class="viewcode-block" id="TomboReads.iter_cov_regs"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.iter_cov_regs">[docs]</a>    <span class="k">def</span> <span class="nf">iter_cov_regs</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">cov_thresh</span><span class="p">,</span> <span class="n">region_size</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Iterate over regions with coverage greater than or equal to cov_thresh.</span>

<span class="sd">        If region_size is provided, regions are rounded to the nearest region_sized windows and only the region start is yielded (e.g. chrm, strand, start). If not provided (chrm, strand, start, end) is yielded.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">def</span> <span class="nf">round_reg_start</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
            <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">region_size</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">x</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">region_size</span><span class="p">)))</span>
        <span class="k">def</span> <span class="nf">round_reg_end</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
            <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">region_size</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">x</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">region_size</span><span class="p">)))</span>


        <span class="k">for</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">starts</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_coverage_regions</span><span class="p">(</span>
                <span class="n">ctrl_reads_index</span><span class="p">):</span>
            <span class="n">curr_reg_start</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="n">valid_cov</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span>
                <span class="p">[</span><span class="kc">False</span><span class="p">,],</span> <span class="n">np</span><span class="o">.</span><span class="n">greater_equal</span><span class="p">(</span><span class="n">cov</span><span class="p">,</span> <span class="n">cov_thresh</span><span class="p">),</span> <span class="p">[</span><span class="kc">False</span><span class="p">,]])))[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">cov_start_i</span><span class="p">,</span> <span class="n">cov_end_i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">valid_cov</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">valid_cov</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
                <span class="n">cov_start</span><span class="p">,</span> <span class="n">cov_end</span> <span class="o">=</span> <span class="n">starts</span><span class="p">[</span><span class="n">cov_start_i</span><span class="p">],</span> <span class="n">starts</span><span class="p">[</span><span class="n">cov_end_i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">region_size</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">yield</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">cov_start</span><span class="p">,</span> <span class="n">cov_end</span>
                    <span class="k">continue</span>
                <span class="k">for</span> <span class="n">reg_start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">round_reg_start</span><span class="p">(</span><span class="n">cov_start</span><span class="p">),</span>
                                       <span class="n">round_reg_end</span><span class="p">(</span><span class="n">cov_end</span><span class="p">),</span> <span class="n">region_size</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">reg_start</span> <span class="o">!=</span> <span class="n">curr_reg_start</span><span class="p">:</span>
                        <span class="k">yield</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">reg_start</span>
                        <span class="n">curr_reg_start</span> <span class="o">=</span> <span class="n">reg_start</span>

        <span class="k">return</span></div>

<div class="viewcode-block" id="TomboReads.get_all_cs"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.get_all_cs">[docs]</a>    <span class="k">def</span> <span class="nf">get_all_cs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get list of all (chromosome, strand) stored in index.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span></div>

<div class="viewcode-block" id="TomboReads.is_empty"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.is_empty">[docs]</a>    <span class="k">def</span> <span class="nf">is_empty</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Are any reads stored in the index?</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">cs_reads</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs_reads</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="kc">True</span></div>

    <span class="k">def</span> <span class="nf">__contains__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm_strand</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">chrm_strand</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span>

    <span class="k">def</span> <span class="nf">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_iter</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">__next__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">next</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_iter</span><span class="p">)</span>

    <span class="c1"># for python2 compatibility</span>
<div class="viewcode-block" id="TomboReads.next"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.next">[docs]</a>    <span class="k">def</span> <span class="nf">next</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="fm">__next__</span><span class="p">()</span></div>

<div class="viewcode-block" id="TomboReads.iter_reads"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.iter_reads">[docs]</a>    <span class="k">def</span> <span class="nf">iter_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Iterate over reads stored in the index</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">cs_reads</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">rd</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">rd</span>
        <span class="k">return</span></div>

<div class="viewcode-block" id="TomboReads.get_cs_reads"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.get_cs_reads">[docs]</a>    <span class="k">def</span> <span class="nf">get_cs_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">invalid_return</span><span class="o">=</span><span class="p">[]):</span>
        <span class="sd">&quot;&quot;&quot;Extract the list of reads stored in a specified chromosome and strand</span>

<span class="sd">        Args:</span>

<span class="sd">            chrm (str): chromosome name</span>
<span class="sd">            strand (str): strand (&#39;+&#39; or &#39;-&#39;)</span>
<span class="sd">            invalid_return: value to return for invalid (chrm, strand)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">invalid_return</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads_index</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span></div>

    <span class="k">def</span> <span class="nf">_get_strand_spec_cov</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">invalid_return</span><span class="p">):</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">invalid_return</span>
        <span class="k">if</span> <span class="n">pos</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">return</span> <span class="n">invalid_return</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)][</span><span class="n">pos</span><span class="p">]</span>

<div class="viewcode-block" id="TomboReads.get_coverage"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.get_coverage">[docs]</a>    <span class="k">def</span> <span class="nf">get_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">invalid_return</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get coverage at specified position</span>

<span class="sd">        Args:</span>
<span class="sd">            chrm (str): chromosome name</span>
<span class="sd">            pos (int): 0-based genomic position</span>
<span class="sd">            strand (str): interval strand (&#39;+&#39;, &#39;-&#39; or None). Default: None (max over both strands)</span>
<span class="sd">            invalid__return: return value for invalid (bad chrm, strand or pos beyond coverage) position. Default: 0</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_compute_coverage</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">max</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_get_strand_spec_cov</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">invalid_return</span><span class="p">),</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_get_strand_spec_cov</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">invalid_return</span><span class="p">))</span>
            <span class="c1"># with None invalid_return value could be max over 2 None&#39;s</span>
            <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">invalid_return</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_strand_spec_cov</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">invalid_return</span><span class="p">)</span></div>

<div class="viewcode-block" id="TomboReads.get_cs_coverage"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.get_cs_coverage">[docs]</a>    <span class="k">def</span> <span class="nf">get_cs_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">invalid_return</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Extract coverage levels over a specified chromosome and strand</span>

<span class="sd">        Args:</span>

<span class="sd">            chrm (str): chromosome name</span>
<span class="sd">            strand (str): strand (&#39;+&#39; or &#39;-&#39;)</span>
<span class="sd">            invalid_return: value to return for invalid (chrm, strand)</span>

<span class="sd">        Returns:</span>
<span class="sd">            numpy array (numpy.int64) containing read coverage levels</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_compute_coverage</span><span class="p">()</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">invald_return</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span></div>

<div class="viewcode-block" id="TomboReads.iter_cs_coverage"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.TomboReads.iter_cs_coverage">[docs]</a>    <span class="k">def</span> <span class="nf">iter_cs_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Iterate over coverage levels across all stored (chrm, strands)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_compute_coverage</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="o">.</span><span class="n">items</span><span class="p">()</span></div></div>


<span class="c1">###########################################</span>
<span class="c1">###### Events Table Access Functions ######</span>
<span class="c1">###########################################</span>

<div class="viewcode-block" id="get_multiple_slots_read_centric"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_multiple_slots_read_centric">[docs]</a><span class="k">def</span> <span class="nf">get_multiple_slots_read_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_names</span><span class="p">,</span> <span class="n">corr_grp</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract multiple read-centric slot_names from this read&#39;s Events table</span>

<span class="sd">    Args:</span>

<span class="sd">        fast5_data (:class:`tombo.tombo_helper.readData` or an open ``h5py.File`` object)</span>
<span class="sd">        slot_name (str): slot from which to extract data (valid values: norm_mean, norm_stdev, start, length, base)</span>
<span class="sd">        corr_grp (str): corrected group slot from which to extract data (default: use value from ``fast5_data`` object; required for ``h5py.File``)</span>

<span class="sd">    Returns:</span>
<span class="sd">        A tuple of numpy arrays specified by the ``slot_names``</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">do_close</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">):</span>
            <span class="n">do_close</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span>
            <span class="n">r_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
        <span class="n">event_slot_name</span> <span class="o">=</span> <span class="s1">&#39;/&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="s1">&#39;Events&#39;</span><span class="p">))</span>
        <span class="c1"># note that it&#39;s more efficient to try to access the slot</span>
        <span class="c1"># and except the error that check if the slot exists first</span>
        <span class="n">r_event_data</span> <span class="o">=</span> <span class="n">r_data</span><span class="p">[</span><span class="n">event_slot_name</span><span class="p">][:]</span>
        <span class="k">if</span> <span class="n">do_close</span><span class="p">:</span> <span class="n">r_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="c1"># probably truncated file or events don&#39;t exist</span>
        <span class="k">return</span> <span class="p">[</span><span class="kc">None</span><span class="p">,]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">slot_names</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">[</span><span class="n">r_event_data</span><span class="p">[</span><span class="n">slot_name</span><span class="p">]</span> <span class="k">for</span> <span class="n">slot_name</span> <span class="ow">in</span> <span class="n">slot_names</span><span class="p">]</span></div>

<div class="viewcode-block" id="get_single_slot_read_centric"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.get_single_slot_read_centric">[docs]</a><span class="k">def</span> <span class="nf">get_single_slot_read_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">,</span> <span class="n">corr_grp</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract read-centric slot_name from this read&#39;s Events table</span>

<span class="sd">    Args:</span>

<span class="sd">        fast5_data (:class:`tombo.tombo_helper.readData` or an open ``h5py.File`` object)</span>
<span class="sd">        slot_name (str): slot from which to extract data (valid values: norm_mean, norm_stdev, start, length, base)</span>
<span class="sd">        corr_grp (str): corrected group slot from which to extract data (default: use value from ``fast5_data`` object; required for ``h5py.File``)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># if r_data is an open h5py object then don&#39;t open the filename</span>
        <span class="n">do_close</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">):</span>
            <span class="n">do_close</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span>
            <span class="n">r_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
        <span class="c1"># note that it&#39;s more efficient to try to access the slot</span>
        <span class="c1"># and except the error that check if the slot exists first</span>
        <span class="n">r_slot_values</span> <span class="o">=</span> <span class="n">r_data</span><span class="p">[</span><span class="s1">&#39;/&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="s1">&#39;Events&#39;</span><span class="p">))][</span>
            <span class="n">slot_name</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">do_close</span><span class="p">:</span> <span class="n">r_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="c1"># probably truncated file or events don&#39;t exist</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="n">r_slot_values</span></div>

<span class="k">def</span> <span class="nf">get_single_slot_genome_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract genome-centric slot_name from this read&#39;s Events table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">r_slot_values</span> <span class="o">=</span> <span class="n">get_single_slot_read_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">r_slot_values</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="n">r_slot_values</span> <span class="o">=</span> <span class="n">r_slot_values</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">r_slot_values</span>

<span class="k">def</span> <span class="nf">get_mean_slot_genome_centric</span><span class="p">(</span><span class="n">cs_reads</span><span class="p">,</span> <span class="n">chrm_len</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get the mean over all reads at each covered genomic location for this slots value</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">base_sums</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">chrm_len</span><span class="p">)</span>
    <span class="n">base_cov</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">chrm_len</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">:</span>
        <span class="c1"># extract read means data so data across all chrms is not</span>
        <span class="c1"># in RAM at one time</span>
        <span class="n">g_slot_values</span> <span class="o">=</span> <span class="n">get_single_slot_genome_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">g_slot_values</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">continue</span>

        <span class="n">base_sums</span><span class="p">[</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
                  <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">g_slot_values</span><span class="p">)]</span> <span class="o">+=</span> <span class="n">g_slot_values</span>
        <span class="n">base_cov</span><span class="p">[</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">g_slot_values</span><span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">return</span> <span class="n">base_sums</span> <span class="o">/</span> <span class="n">base_cov</span>

<span class="k">def</span> <span class="nf">iter_mean_slot_values</span><span class="p">(</span>
        <span class="n">reads_index</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">,</span> <span class="n">slot_name</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Iterate through chromosomes and strands yielding mean slots values over</span>
<span class="sd">    all reads at each covered genomic location.</span>

<span class="sd">    Generator returns chrmosome, strand, cs_mean_values tuples (3 return values).</span>

<span class="sd">    If a second raw_read_coverage object is included another cs_mean_values</span>
<span class="sd">    array is yeilded (4 return values)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># ignore divide by zero errors that occur where there is no</span>
    <span class="c1"># coverage. Need to correct nan values after subtracting two sets of</span>
    <span class="c1"># coverage so leave as nan for now</span>
    <span class="n">old_err_settings</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="nb">all</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span> <span class="ow">in</span> <span class="p">[(</span><span class="n">c</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">chrm_sizes</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">)]:</span>
        <span class="k">if</span> <span class="n">ctrl_reads_index</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">reads_index</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">cs_mean_values</span> <span class="o">=</span> <span class="n">get_mean_slot_genome_centric</span><span class="p">(</span>
                <span class="n">reads_index</span><span class="o">.</span><span class="n">get_cs_reads</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span>
                <span class="n">chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">],</span> <span class="n">slot_name</span><span class="p">)</span>
            <span class="k">yield</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">cs_mean_values</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cs_mean_values</span><span class="p">,</span> <span class="n">ctrl_cs_mean_values</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">in</span> <span class="n">reads_index</span><span class="p">:</span>
                <span class="n">cs_mean_values</span> <span class="o">=</span> <span class="n">get_mean_slot_genome_centric</span><span class="p">(</span>
                    <span class="n">reads_index</span><span class="o">.</span><span class="n">get_cs_reads</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span>
                    <span class="n">chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">],</span> <span class="n">slot_name</span><span class="p">)</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="ow">in</span> <span class="n">ctrl_reads_index</span><span class="p">:</span>
                <span class="n">ctrl_cs_mean_values</span> <span class="o">=</span> <span class="n">get_mean_slot_genome_centric</span><span class="p">(</span>
                    <span class="n">ctrl_reads_index</span><span class="o">.</span><span class="n">get_cs_reads</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span>
                    <span class="n">chrm_sizes</span><span class="p">[</span><span class="n">chrm</span><span class="p">],</span> <span class="n">slot_name</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">cs_mean_values</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">ctrl_cs_mean_values</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">continue</span>
            <span class="k">yield</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">cs_mean_values</span><span class="p">,</span> <span class="n">ctrl_cs_mean_values</span>

    <span class="n">_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="o">**</span><span class="n">old_err_settings</span><span class="p">)</span>

    <span class="k">return</span>

<span class="k">def</span> <span class="nf">get_largest_signal_differences</span><span class="p">(</span>
        <span class="n">reads_index</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="p">,</span> <span class="n">num_regions</span><span class="p">,</span> <span class="n">num_bases</span><span class="p">):</span>
    <span class="n">chrm_sizes</span> <span class="o">=</span> <span class="n">get_chrm_sizes</span><span class="p">(</span><span class="n">reads_index</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="p">)</span>

    <span class="n">all_largest_diff_poss</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">cs_sig_means</span><span class="p">,</span> <span class="n">ctrl_cs_sig_means</span> <span class="ow">in</span> <span class="n">iter_mean_slot_values</span><span class="p">(</span>
            <span class="n">reads_index</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">,</span> <span class="s1">&#39;norm_mean&#39;</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">cs_sig_means</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ctrl_cs_sig_means</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">continue</span>
        <span class="n">chrm_diffs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">cs_sig_means</span> <span class="o">-</span> <span class="n">ctrl_cs_sig_means</span><span class="p">))</span>
        <span class="n">chrm_max_diff_regs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">chrm_diffs</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="n">num_regions</span><span class="p">]</span>
        <span class="n">all_largest_diff_poss</span><span class="o">.</span><span class="n">extend</span><span class="p">((</span>
            <span class="n">chrm_diffs</span><span class="p">[</span><span class="n">pos</span><span class="p">],</span> <span class="nb">max</span><span class="p">(</span><span class="n">pos</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">num_bases</span> <span class="o">/</span> <span class="mf">2.0</span><span class="p">),</span> <span class="mi">0</span><span class="p">),</span>
            <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span> <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">chrm_max_diff_regs</span><span class="p">)</span>

    <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_largest_diff_poss</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)[:</span><span class="n">num_regions</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">get_signal_differences</span><span class="p">(</span><span class="n">reads_index</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Helper function to compute all signal differences</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">chrm_sizes</span> <span class="o">=</span> <span class="n">get_chrm_sizes</span><span class="p">(</span><span class="n">reads_index</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="p">)</span>

    <span class="n">all_diffs</span> <span class="o">=</span>  <span class="p">{}</span>
    <span class="k">for</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">cs_sig_means</span><span class="p">,</span> <span class="n">ctrl_cs_sig_means</span> <span class="ow">in</span> <span class="n">iter_mean_slot_values</span><span class="p">(</span>
            <span class="n">reads_index</span><span class="p">,</span> <span class="n">chrm_sizes</span><span class="p">,</span> <span class="s1">&#39;norm_mean&#39;</span><span class="p">,</span> <span class="n">ctrl_reads_index</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">cs_sig_means</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ctrl_cs_sig_means</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">continue</span>
        <span class="n">all_diffs</span><span class="p">[(</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span>
            <span class="n">cs_sig_means</span> <span class="o">-</span> <span class="n">ctrl_cs_sig_means</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">all_diffs</span>


<span class="c1">####################################</span>
<span class="c1">###### Genomic Interval Class ######</span>
<span class="c1">####################################</span>

<div class="viewcode-block" id="intervalData"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData">[docs]</a><span class="k">class</span> <span class="nc">intervalData</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Genome/transcriptome interval information</span>

<span class="sd">    Example::</span>

<span class="sd">        int_data = tombo_helper.intervalData(</span>
<span class="sd">            chrm=&#39;chr20&#39;, start=10000, end=10100, strand=&#39;+&#39;)</span>

<span class="sd">    Note:</span>
<span class="sd">        All intervalData functions return the object in order to allow chaining of interval commands. (e.g. ``int_data.add_reads(reads_index).add_seq()``)</span>

<span class="sd">    .. automethod:: __init__</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__setattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s1">&#39;chrm&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">unicode</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="s1">&#39; must be a string&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">integer</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="s1">&#39; must be an int&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s1">&#39;strand&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">value</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">&#39;strand must be either None, &quot;+&quot;, or &quot;-&quot;&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;reg_id&#39;</span><span class="p">,</span> <span class="s1">&#39;reg_text&#39;</span><span class="p">,</span> <span class="s1">&#39;seq&#39;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">unicode</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="s1">&#39; must be a string&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s1">&#39;reads&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">&#39;reads must be a list&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="s1">&#39; is an invalid attribute for intervalData&#39;</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">intervalData</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__setattr__</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>

<div class="viewcode-block" id="intervalData.__init__"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrm</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reg_id</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">reg_text</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">reads</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seq</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Initialize a genome/transcriptome interval object</span>

<span class="sd">        Args:</span>
<span class="sd">            chrm (str): chromosome name.</span>
<span class="sd">            start (int): 0-based start position.</span>
<span class="sd">            end (int): 1-based (or open interval) end position.</span>
<span class="sd">            strand (str): interval strand (&#39;+&#39;, &#39;-&#39; or None). Default: None</span>
<span class="sd">            reg_id (str): used to keep track of multiple intervals. Default: None</span>
<span class="sd">            reg_text (str): region description text. Default: &#39;&#39;</span>
<span class="sd">            reads (list): list of readData values overlapping region. Default: None</span>
<span class="sd">            seq (str): region genomic sequence. Default: None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chrm</span> <span class="o">=</span> <span class="n">unicode</span><span class="p">(</span><span class="n">chrm</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">end</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="n">unicode</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span> <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reg_id</span> <span class="o">=</span> <span class="n">unicode</span><span class="p">(</span><span class="n">reg_id</span><span class="p">)</span> <span class="k">if</span> <span class="n">reg_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reg_text</span> <span class="o">=</span> <span class="n">unicode</span><span class="p">(</span><span class="n">reg_text</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads</span> <span class="o">=</span> <span class="n">reads</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span></div>

<div class="viewcode-block" id="intervalData.update"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData.update">[docs]</a>    <span class="k">def</span> <span class="nf">update</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Update slots specified in keyword arguments (kwargs)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="fm">__setattr__</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span>
        <span class="c1"># return self to allow chaining and auto-return</span>
        <span class="k">return</span> <span class="bp">self</span></div>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># convert reads and seq if they are too long</span>
        <span class="n">self_vars</span> <span class="o">=</span> <span class="nb">vars</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;reads&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;reads&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&lt;</span><span class="si">{:d}</span><span class="s1"> reads&gt;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;reads&#39;</span><span class="p">]))</span>
        <span class="k">if</span> <span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;seq&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;seq&#39;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">50</span><span class="p">:</span>
            <span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;seq&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&lt;</span><span class="si">{:d}</span><span class="s1"> bases of sequence&gt;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;seq&#39;</span><span class="p">]))</span>
        <span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;reg_text&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&quot;&#39;</span> <span class="o">+</span> <span class="n">self_vars</span><span class="p">[</span><span class="s1">&#39;reg_text&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;&quot;&#39;</span>
        <span class="k">return</span> <span class="s1">&#39;&lt;tombo.tombo_helper.intervalData object&gt; :</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{:&gt;15}</span><span class="s2"> : </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">v</span><span class="p">))</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">self_vars</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

<div class="viewcode-block" id="intervalData.copy"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData.copy">[docs]</a>    <span class="k">def</span> <span class="nf">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include_reads</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create a copy of this interval. Useful when adding sets of reads from multiple samples to compare.</span>

<span class="sd">        Args:</span>
<span class="sd">            include_reads (bool): include current reads slot in the new object (default: True)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cp_reads</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads</span> <span class="k">if</span> <span class="n">include_reads</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                          <span class="bp">self</span><span class="o">.</span><span class="n">reg_id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">reg_text</span><span class="p">,</span> <span class="n">cp_reads</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span></div>

<div class="viewcode-block" id="intervalData.merge"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData.merge">[docs]</a>    <span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other_reg</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create a copy of this interval with the reads from this interval and `other_reg`</span>

<span class="sd">        Args:</span>
<span class="sd">            other_reg (:class:`tombo.tombo_helper.intervalData`): a second region to merge with this interval</span>

<span class="sd">        Note:</span>
<span class="sd">            Aside from reads, all other attributes will be taken from this interval</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">merged_reg_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">self_reads</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="p">[]</span>
        <span class="n">other_reads</span> <span class="o">=</span> <span class="n">other_reg</span><span class="o">.</span><span class="n">reads</span> <span class="k">if</span> <span class="n">other_reg</span><span class="o">.</span><span class="n">reads</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="p">[]</span>
        <span class="k">return</span> <span class="n">merged_reg_data</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">reads</span><span class="o">=</span><span class="n">self_reads</span> <span class="o">+</span> <span class="n">other_reads</span><span class="p">)</span></div>

<div class="viewcode-block" id="intervalData.expand_interval"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData.expand_interval">[docs]</a>    <span class="k">def</span> <span class="nf">expand_interval</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">expand_width</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Expand this interval by the specified amount (effects only the start and stop attributes; NOT seq or reads)</span>

<span class="sd">        Args:</span>

<span class="sd">            expand_width (int): amount by which to expand the interval</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">expand_width</span><span class="p">,</span>
                    <span class="n">end</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">+</span> <span class="n">expand_width</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span></div>

<div class="viewcode-block" id="intervalData.add_reads"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData.add_reads">[docs]</a>    <span class="k">def</span> <span class="nf">add_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reads_index</span><span class="p">,</span> <span class="n">require_full_span</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add reads overlapping  this interval</span>

<span class="sd">        Args:</span>
<span class="sd">            reads_index (:class:`tombo.tombo_helper.TomboReads`): reads index</span>
<span class="sd">            require_full_span (bool): require that reads span then entire interval (default: include all reads overlapping the interval)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">def</span> <span class="nf">get_c_s_data</span><span class="p">(</span><span class="n">strand</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">require_full_span</span><span class="p">:</span>
                <span class="k">def</span> <span class="nf">read_intersects_interval</span><span class="p">(</span><span class="n">r_start</span><span class="p">,</span> <span class="n">r_end</span><span class="p">):</span>
                    <span class="k">return</span> <span class="n">r_start</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="ow">and</span> <span class="n">r_end</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">def</span> <span class="nf">read_intersects_interval</span><span class="p">(</span><span class="n">r_start</span><span class="p">,</span> <span class="n">r_end</span><span class="p">):</span>
                    <span class="k">return</span> <span class="ow">not</span> <span class="p">(</span><span class="n">r_start</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="ow">or</span> <span class="n">r_end</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>

            <span class="c1"># get all reads intersecting the interval</span>
            <span class="k">return</span> <span class="p">[</span>
                <span class="n">r_data</span> <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">reads_index</span><span class="o">.</span><span class="n">get_cs_reads</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">read_intersects_interval</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span><span class="p">)]</span>


        <span class="c1"># get all reads that overlap this interval</span>
        <span class="c1"># note that this includes partial overlaps as these contribute</span>
        <span class="c1"># to coverage and other statistics so can&#39;t really restrict to</span>
        <span class="c1"># full coverage as previous versions of code did</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">reads</span><span class="o">=</span><span class="n">get_c_s_data</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">strand</span><span class="p">))</span>

        <span class="c1"># if strand is None, get data from both strands</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">reads</span><span class="o">=</span><span class="n">get_c_s_data</span><span class="p">(</span><span class="s1">&#39;+&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="n">get_c_s_data</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">))</span></div>

    <span class="k">def</span> <span class="nf">_update_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">r_data</span><span class="p">,</span> <span class="n">reg_base_data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Update the sequence for the region based on this read</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">read_bases</span> <span class="o">=</span> <span class="n">get_single_slot_read_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="s1">&#39;base&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">read_bases</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">warning_message</span><span class="p">(</span>
                <span class="s1">&#39;Unable to extract data from read. Potentially corrupted file &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;or invalid Tombo index file for this directory.&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
        <span class="n">r_seq</span> <span class="o">=</span> <span class="sa">b</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">read_bases</span><span class="p">)</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
            <span class="n">r_seq</span> <span class="o">=</span> <span class="n">rev_comp</span><span class="p">(</span><span class="n">r_seq</span><span class="p">)</span>

        <span class="c1"># if read starts before the interval</span>
        <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
            <span class="n">r_end_overlap</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span>
            <span class="c1"># if read covers the whole interval</span>
            <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">:</span>
                <span class="n">r_end_clip</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span>
                <span class="n">reg_base_data</span> <span class="o">=</span> <span class="n">r_seq</span><span class="p">[</span><span class="o">-</span><span class="n">r_end_overlap</span><span class="p">:</span><span class="o">-</span><span class="n">r_end_clip</span><span class="p">]</span>
                <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">)</span>
            <span class="c1"># end of read overlaps beginning of interval</span>
            <span class="n">reg_base_data</span><span class="p">[:</span><span class="n">r_end_overlap</span><span class="p">]</span> <span class="o">=</span> <span class="n">r_seq</span><span class="p">[</span><span class="o">-</span><span class="n">r_end_overlap</span><span class="p">:]</span>
            <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">r_end_overlap</span>
        <span class="c1"># read doesn&#39;t cover the beginning of region</span>
        <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">:</span>
            <span class="c1"># beginning of read covers to the end of the region</span>
            <span class="n">r_begin_overlap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span>
            <span class="n">reg_base_data</span><span class="p">[</span><span class="o">-</span><span class="n">r_begin_overlap</span><span class="p">:]</span> <span class="o">=</span> <span class="n">r_seq</span><span class="p">[:</span><span class="n">r_begin_overlap</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">)</span>
        <span class="c1"># first read is completely contained in the interval</span>
        <span class="n">r_len</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span>
        <span class="n">r_int_start</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span>
        <span class="n">reg_base_data</span><span class="p">[</span><span class="n">r_int_start</span><span class="p">:</span><span class="n">r_int_start</span> <span class="o">+</span> <span class="n">r_len</span><span class="p">]</span> <span class="o">=</span> <span class="n">r_seq</span>
        <span class="k">return</span> <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">r_int_start</span> <span class="o">+</span> <span class="n">r_len</span>

<div class="viewcode-block" id="intervalData.add_seq"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData.add_seq">[docs]</a>    <span class="k">def</span> <span class="nf">add_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome_index</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Extract the forward strand genomic sequence for an interval from reads or genome_index if provided</span>

<span class="sd">        Args:</span>
<span class="sd">            genome_index (:class:`tombo.tombo_helper.Fasta`) Tombo FASTA sequence object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">genome_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="n">genome_index</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">))</span>

        <span class="c1"># handle case where no read overlaps whole region</span>
        <span class="c1"># let each read contibute its sequence and fill the rest</span>
        <span class="c1"># with dashes</span>
        <span class="n">reg_base_data</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;-&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">))</span>
        <span class="c1"># get region sequence by moving through reads that</span>
        <span class="c1"># cover the region, but only extract seqeunce from the</span>
        <span class="c1"># (close to) minimal reads</span>
        <span class="n">s_reg_reads</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">r</span><span class="p">:</span> <span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">r</span><span class="o">.</span><span class="n">end</span><span class="p">))</span>
        <span class="c1"># begin filling sequence with first (by start pos) read</span>
        <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">curr_cov_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_update_seq</span><span class="p">(</span>
            <span class="n">s_reg_reads</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">reg_base_data</span><span class="p">)</span>
        <span class="c1"># if there was only one read return now</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">s_reg_reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">curr_cov_pos</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">))</span>

        <span class="c1"># get next read (by start pos)</span>
        <span class="n">curr_read</span> <span class="o">=</span> <span class="n">s_reg_reads</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">next_read</span> <span class="ow">in</span> <span class="n">s_reg_reads</span><span class="p">:</span>
            <span class="c1"># once the next read start passes the region covered thus far</span>
            <span class="c1"># add the sequence for the saved curr_read to the reg sequence</span>
            <span class="k">if</span> <span class="n">next_read</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;=</span> <span class="n">curr_cov_pos</span><span class="p">:</span>
                <span class="c1"># add read with curr longest end position to the region seq</span>
                <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">curr_cov_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_update_seq</span><span class="p">(</span>
                    <span class="n">curr_read</span><span class="p">,</span> <span class="n">reg_base_data</span><span class="p">)</span>
                <span class="n">curr_read</span> <span class="o">=</span> <span class="n">next_read</span>
                <span class="c1"># if the whole interval is covered return the sequence</span>
                <span class="k">if</span> <span class="n">curr_cov_pos</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
                    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">))</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">next_read</span><span class="o">.</span><span class="n">end</span> <span class="o">&gt;</span> <span class="n">curr_read</span><span class="o">.</span><span class="n">end</span><span class="p">:</span>
                <span class="n">curr_read</span> <span class="o">=</span> <span class="n">next_read</span>

        <span class="n">reg_base_data</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_update_seq</span><span class="p">(</span><span class="n">curr_read</span><span class="p">,</span> <span class="n">reg_base_data</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reg_base_data</span><span class="p">))</span></div>

<div class="viewcode-block" id="intervalData.get_base_levels"><a class="viewcode-back" href="../../tombo.html#tombo.tombo_helper.intervalData.get_base_levels">[docs]</a>    <span class="k">def</span> <span class="nf">get_base_levels</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">read_rows</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">num_reads</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Extract base levels from this interval.</span>

<span class="sd">        Args:</span>
<span class="sd">            read_rows (bool): return array with reads as rows (default: row columns)</span>
<span class="sd">            num_reads (int): maximal number of reads to output. randomly downsample if more reads are present (default: output all reads)</span>

<span class="sd">        Return:</span>
<span class="sd">            `np.array` containing read mean levels with rows corresponding to interval position and columns to individual reads (or reverse if `read_rows`)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">def</span> <span class="nf">get_read_reg_events</span><span class="p">(</span><span class="n">r_data</span><span class="p">):</span>
            <span class="n">r_means</span> <span class="o">=</span> <span class="n">get_single_slot_genome_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="s1">&#39;norm_mean&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">r_means</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">return</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="ow">and</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">:</span>
                <span class="c1"># handle reads that are contained in a region</span>
                <span class="n">start_overlap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span>
                <span class="n">end_overlap</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span>
                <span class="c1"># create region with nan values</span>
                <span class="n">r_reg_means</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="n">r_reg_means</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NAN</span>
                <span class="n">r_reg_means</span><span class="p">[</span><span class="o">-</span><span class="n">start_overlap</span><span class="p">:</span><span class="n">end_overlap</span><span class="p">]</span> <span class="o">=</span> <span class="n">r_means</span><span class="p">[</span>
                    <span class="o">-</span><span class="n">end_overlap</span><span class="p">:</span><span class="n">start_overlap</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
                <span class="c1"># handle reads that start in middle of region</span>
                <span class="n">start_overlap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span>
                <span class="c1"># create region with nan values</span>
                <span class="n">r_reg_means</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="n">r_reg_means</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NAN</span>
                <span class="n">r_reg_means</span><span class="p">[</span><span class="o">-</span><span class="n">start_overlap</span><span class="p">:]</span> <span class="o">=</span> <span class="n">r_means</span><span class="p">[:</span><span class="n">start_overlap</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">:</span>
                <span class="c1"># handle reads that end inside region</span>
                <span class="n">end_overlap</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span>
                <span class="c1"># create region with nan values</span>
                <span class="n">r_reg_means</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="n">r_reg_means</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NAN</span>
                <span class="n">r_reg_means</span><span class="p">[:</span><span class="n">end_overlap</span><span class="p">]</span> <span class="o">=</span> <span class="n">r_means</span><span class="p">[</span><span class="o">-</span><span class="n">end_overlap</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">r_reg_means</span> <span class="o">=</span> <span class="n">r_means</span><span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">]</span>

            <span class="k">return</span> <span class="n">r_reg_means</span>


        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span>
                <span class="s1">&#39;Must annotate region with reads &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;(see `TomboInterval.add_reads`) to extract base levels.&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_reads</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span>
        <span class="n">reg_events</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">r_means</span> <span class="o">=</span> <span class="n">get_read_reg_events</span><span class="p">(</span><span class="n">r_data</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">r_means</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">reg_events</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r_means</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">num_reads</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_events</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">num_reads</span><span class="p">:</span>
                <span class="k">break</span>

        <span class="k">if</span> <span class="n">read_rows</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">row_stack</span><span class="p">(</span><span class="n">reg_events</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">(</span><span class="n">reg_events</span><span class="p">)</span></div></div>


<span class="k">def</span> <span class="nf">filter_empty_regions</span><span class="p">(</span><span class="n">plot_intervals</span><span class="p">):</span>
    <span class="n">num_filt</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p_int</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">p_int</span> <span class="ow">in</span> <span class="n">plot_intervals</span><span class="p">)</span>
    <span class="n">plot_intervals</span> <span class="o">=</span> <span class="p">[</span><span class="n">p_int</span> <span class="k">for</span> <span class="n">p_int</span> <span class="ow">in</span> <span class="n">plot_intervals</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">p_int</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">plot_intervals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">error_message_and_exit</span><span class="p">(</span><span class="s1">&#39;No reads in any selected regions.&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">VERBOSE</span> <span class="ow">and</span> <span class="n">num_filt</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">warning_message</span><span class="p">(</span><span class="s1">&#39;Some selected regions contain no reads.&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">plot_intervals</span>

<span class="k">def</span> <span class="nf">get_unique_intervals</span><span class="p">(</span><span class="n">plot_intervals</span><span class="p">,</span> <span class="n">covered_poss</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">num_regions</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="c1"># unique genomic regions filter</span>
    <span class="n">uniq_p_intervals</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">used_intervals</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">reg_data</span> <span class="ow">in</span> <span class="n">plot_intervals</span><span class="p">:</span>
        <span class="c1"># could have significant region immediately next to</span>
        <span class="c1"># beginning/end of reads</span>
        <span class="n">interval_poss</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">reg_data</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">reg_data</span><span class="o">.</span><span class="n">end</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">reg_data</span><span class="o">.</span><span class="n">start</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">used_intervals</span><span class="p">[(</span>
                <span class="n">reg_data</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="n">reg_data</span><span class="o">.</span><span class="n">strand</span><span class="p">)]</span> <span class="ow">and</span> <span class="p">(</span>
                    <span class="n">covered_poss</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="nb">all</span><span class="p">(</span>
                        <span class="n">pos</span> <span class="ow">in</span> <span class="n">covered_poss</span><span class="p">[(</span><span class="n">reg_data</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="n">reg_data</span><span class="o">.</span><span class="n">strand</span><span class="p">)]</span>
                        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">interval_poss</span><span class="p">)):</span>
            <span class="n">uniq_p_intervals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reg_data</span><span class="p">)</span>
            <span class="n">used_intervals</span><span class="p">[(</span><span class="n">reg_data</span><span class="o">.</span><span class="n">chrm</span><span class="p">,</span> <span class="n">reg_data</span><span class="o">.</span><span class="n">strand</span><span class="p">)]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="n">interval_poss</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_regions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">uniq_p_intervals</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">num_regions</span><span class="p">:</span>
            <span class="k">break</span>

    <span class="k">return</span> <span class="n">uniq_p_intervals</span>


<span class="c1">###########################################</span>
<span class="c1">###### Special Data Access Functions ######</span>
<span class="c1">###########################################</span>

<span class="k">def</span> <span class="nf">get_channel_info</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get channel information for a read</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">fast5_attrs</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;UniqueGlobalKey/channel_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s2">&quot;No channel_id group in HDF5 file. &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;Probably mux scan HDF5 file.&quot;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">channel_info</span> <span class="o">=</span> <span class="n">channelInfo</span><span class="p">(</span>
            <span class="n">fast5_attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;offset&#39;</span><span class="p">),</span> <span class="n">fast5_attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;range&#39;</span><span class="p">),</span>
            <span class="n">fast5_attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;digitisation&#39;</span><span class="p">),</span> <span class="n">fast5_attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;channel_number&#39;</span><span class="p">),</span>
            <span class="n">fast5_attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;sampling_rate&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">))</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s2">&quot;Channel info parameters not available.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">channel_info</span>

<span class="k">def</span> <span class="nf">get_raw_signal</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="n">int_start</span><span class="p">,</span> <span class="n">int_end</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract raw signal from where this read overlaps a particular genomic region</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
        <span class="c1"># retrieve shift and scale computed in correction script</span>
        <span class="n">corr_subgrp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span><span class="p">]</span>
        <span class="n">event_starts</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
        <span class="n">events_end</span> <span class="o">=</span> <span class="n">event_starts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">corr_subgrp</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">segs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">event_starts</span><span class="p">,</span> <span class="p">[</span><span class="n">events_end</span><span class="p">,]])</span>

        <span class="n">scale_values</span> <span class="o">=</span> <span class="n">scaleValues</span><span class="p">(</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;shift&#39;</span><span class="p">),</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;scale&#39;</span><span class="p">),</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;lower_lim&#39;</span><span class="p">),</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;upper_lim&#39;</span><span class="p">),</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;outlier_threshold&#39;</span><span class="p">))</span>
        <span class="n">all_sig</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)[</span><span class="s1">&#39;Signal&#39;</span><span class="p">][:]</span>

    <span class="n">rsrtr</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">read_start_rel_to_raw</span>
    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">rna</span><span class="p">:</span>
        <span class="c1"># reverse raw signal for RNA</span>
        <span class="n">all_sig</span> <span class="o">=</span> <span class="n">all_sig</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
        <span class="n">segs</span> <span class="o">=</span> <span class="p">(</span><span class="n">segs</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">segs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">int_start</span> <span class="o">&lt;</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
        <span class="c1"># handle reads that start in the middle of the interval</span>
        <span class="n">start_offset</span> <span class="o">=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">int_start</span>
        <span class="n">overlap_seg_data</span> <span class="o">=</span> <span class="n">segs</span><span class="p">[:</span><span class="n">int_end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">start_offset</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">skipped_bases</span> <span class="o">=</span> <span class="n">int_start</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span>
        <span class="n">overlap_seg_data</span> <span class="o">=</span> <span class="n">segs</span><span class="p">[</span>
            <span class="n">skipped_bases</span><span class="p">:</span><span class="n">int_end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>

    <span class="c1"># trim and flip raw signal (perform normalization outside of this</span>
    <span class="c1"># function in order to avoid circular import between tombo_helper</span>
    <span class="c1"># and tombo_stats)</span>
    <span class="n">num_reg_obs</span> <span class="o">=</span> <span class="n">overlap_seg_data</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">overlap_seg_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
        <span class="n">reg_start_rel_raw</span> <span class="o">=</span> <span class="n">rsrtr</span> <span class="o">+</span> <span class="n">overlap_seg_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">r_sig</span> <span class="o">=</span> <span class="n">all_sig</span><span class="p">[</span><span class="n">reg_start_rel_raw</span><span class="p">:</span><span class="n">reg_start_rel_raw</span> <span class="o">+</span> <span class="n">num_reg_obs</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">reg_start_rel_raw</span> <span class="o">=</span> <span class="n">rsrtr</span> <span class="o">+</span> <span class="n">segs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">overlap_seg_data</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">r_sig</span> <span class="o">=</span> <span class="n">all_sig</span><span class="p">[</span><span class="n">reg_start_rel_raw</span><span class="p">:</span><span class="n">reg_start_rel_raw</span> <span class="o">+</span> <span class="n">num_reg_obs</span><span class="p">]</span>
        <span class="n">r_sig</span> <span class="o">=</span> <span class="n">r_sig</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">r_sig</span><span class="p">,</span> <span class="n">overlap_seg_data</span><span class="p">,</span> <span class="n">start_offset</span><span class="p">,</span> <span class="n">scale_values</span>

<span class="k">def</span> <span class="nf">parse_read_correction_data</span><span class="p">(</span><span class="n">r_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parse correction data from an event resquiggled read</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
            <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span><span class="p">]</span>
            <span class="n">events_grp</span> <span class="o">=</span> <span class="n">corr_grp</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">]</span>
            <span class="n">event_starts</span> <span class="o">=</span> <span class="n">events_grp</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
            <span class="n">events_end</span> <span class="o">=</span> <span class="n">event_starts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">events_grp</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">new_segs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">event_starts</span><span class="p">,</span> <span class="p">[</span><span class="n">events_end</span><span class="p">,]])</span>

            <span class="n">raw_grp</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)</span>
            <span class="n">read_id</span> <span class="o">=</span> <span class="n">raw_grp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;read_id&#39;</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">read_id</span> <span class="o">=</span> <span class="n">read_id</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
            <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
                <span class="k">pass</span>
            <span class="n">signal_data</span> <span class="o">=</span> <span class="n">raw_grp</span><span class="p">[</span><span class="s1">&#39;Signal&#39;</span><span class="p">][:]</span>

            <span class="n">raw_offset</span> <span class="o">=</span> <span class="n">events_grp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">)</span>
            <span class="n">scale_values</span> <span class="o">=</span> <span class="n">scaleValues</span><span class="p">(</span><span class="o">*</span><span class="p">[</span>
                <span class="n">corr_grp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">attr_name</span><span class="p">)</span> <span class="k">for</span> <span class="n">attr_name</span> <span class="ow">in</span> <span class="p">(</span>
                    <span class="s1">&#39;shift&#39;</span><span class="p">,</span> <span class="s1">&#39;scale&#39;</span><span class="p">,</span> <span class="s1">&#39;lower_lim&#39;</span><span class="p">,</span> <span class="s1">&#39;upper_lim&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;outlier_threshold&#39;</span><span class="p">)])</span>

            <span class="n">old_segs</span> <span class="o">=</span> <span class="n">corr_grp</span><span class="p">[</span><span class="s1">&#39;Alignment/read_segments&#39;</span><span class="p">][:]</span>
            <span class="n">old_align_vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(),</span>
                <span class="n">corr_grp</span><span class="p">[</span><span class="s1">&#39;Alignment/read_alignment&#39;</span><span class="p">][:]))</span>
            <span class="n">new_align_vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(),</span>
                <span class="n">corr_grp</span><span class="p">[</span><span class="s1">&#39;Alignment/genome_alignment&#39;</span><span class="p">][:]))</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">rna</span><span class="p">:</span>
        <span class="n">signal_data</span> <span class="o">=</span> <span class="n">signal_data</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">read_id</span><span class="p">,</span> <span class="n">signal_data</span><span class="p">,</span> <span class="n">raw_offset</span><span class="p">,</span> <span class="n">scale_values</span><span class="p">,</span> <span class="n">old_segs</span><span class="p">,</span>
            <span class="n">old_align_vals</span><span class="p">,</span> <span class="n">new_align_vals</span><span class="p">,</span> <span class="n">events_end</span><span class="p">,</span> <span class="n">new_segs</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">get_all_read_data</span><span class="p">(</span><span class="n">r_data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract most relevant read data from this read</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">r_data</span><span class="o">.</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
            <span class="c1"># note that it&#39;s more efficient to try to access the slot</span>
            <span class="c1"># and except the error that check if the slot exists first</span>
            <span class="n">corr_subgrp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses/&#39;</span> <span class="o">+</span> <span class="n">r_data</span><span class="o">.</span><span class="n">corr_group</span><span class="p">]</span>
            <span class="n">algn_subgrp</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">corr_subgrp</span><span class="p">[</span><span class="s1">&#39;Alignment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
            <span class="n">event_data</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="p">[</span><span class="s1">&#39;Events&#39;</span><span class="p">][:]</span>
            <span class="n">r_attrs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
            <span class="n">all_sig</span> <span class="o">=</span> <span class="n">get_raw_read_slot</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">)[</span><span class="s1">&#39;Signal&#39;</span><span class="p">][:]</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="c1"># probably truncated file or Events slot doesn&#39;t exist</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">r_data</span><span class="o">.</span><span class="n">rna</span><span class="p">:</span>
        <span class="n">all_sig</span> <span class="o">=</span> <span class="n">all_sig</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">r_means</span> <span class="o">=</span> <span class="n">event_data</span><span class="p">[</span><span class="s1">&#39;norm_mean&#39;</span><span class="p">]</span>
    <span class="n">r_seq</span> <span class="o">=</span> <span class="sa">b</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">event_data</span><span class="p">[</span><span class="s1">&#39;base&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>

    <span class="n">events_end</span> <span class="o">=</span> <span class="n">event_data</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">event_data</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">]</span>
    <span class="n">segs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">event_data</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">events_end</span><span class="p">,]])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">r_means</span><span class="p">,</span> <span class="n">r_seq</span><span class="p">,</span> <span class="n">all_sig</span><span class="p">,</span> <span class="n">segs</span><span class="p">,</span> <span class="n">r_data</span><span class="o">.</span><span class="n">read_start_rel_to_raw</span><span class="p">,</span>
            <span class="n">r_attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;norm_type&#39;</span><span class="p">),</span> <span class="n">r_attrs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;outlier_threshold&#39;</span><span class="p">),</span>
            <span class="n">genomeLocation</span><span class="p">(</span>
                <span class="n">algn_subgrp</span><span class="p">[</span><span class="s1">&#39;mapped_start&#39;</span><span class="p">],</span> <span class="n">algn_subgrp</span><span class="p">[</span><span class="s1">&#39;mapped_strand&#39;</span><span class="p">],</span>
                <span class="n">algn_subgrp</span><span class="p">[</span><span class="s1">&#39;mapped_chrom&#39;</span><span class="p">]))</span>

<span class="k">def</span> <span class="nf">get_reads_events</span><span class="p">(</span><span class="n">cs_reads</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract read base levels split by genomic position</span>

<span class="sd">    Args:</span>

<span class="sd">        cs_reads (list): a list of reads from a single (chromosome, strand)</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dictionary with 0-based genomic positions pointing to read signal levels</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># note that this function assumes that all reads come from the same</span>
    <span class="c1"># chromosome and strand</span>
    <span class="n">cs_base_means</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">cs_read_start_ends</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">r_data</span> <span class="ow">in</span> <span class="n">cs_reads</span><span class="p">:</span>
        <span class="c1"># extract read means data so data across all chrms is not</span>
        <span class="c1"># in RAM at one time</span>
        <span class="n">read_means</span> <span class="o">=</span> <span class="n">get_single_slot_genome_centric</span><span class="p">(</span><span class="n">r_data</span><span class="p">,</span> <span class="s1">&#39;norm_mean&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">read_means</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">continue</span>
        <span class="k">if</span> <span class="n">read_means</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">:</span> <span class="k">continue</span>
        <span class="k">assert</span> <span class="n">read_means</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="p">(</span>
            <span class="s1">&#39;Read found with mismatching mapping location and &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;signal information.&#39;</span><span class="p">)</span>
        <span class="n">cs_base_means</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read_means</span><span class="p">)</span>
        <span class="n">cs_read_start_ends</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">r_data</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">r_data</span><span class="o">.</span><span class="n">end</span><span class="p">))</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs_base_means</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="k">return</span> <span class="kc">None</span>

    <span class="n">chrm_signal</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">cs_base_means</span><span class="p">)</span>
    <span class="n">chrm_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
                               <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="n">cs_read_start_ends</span><span class="p">])</span>
    <span class="c1"># get order of all bases from position array</span>
    <span class="n">as_chrm_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">chrm_pos</span><span class="p">)</span>
    <span class="c1"># then sort the signal array by genomic position and</span>
    <span class="c1"># split into event means by base</span>
    <span class="n">split_poss</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([[</span><span class="mi">0</span><span class="p">,],</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span>
            <span class="n">chrm_pos</span><span class="p">[</span><span class="n">as_chrm_pos</span><span class="p">])])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">cs_base_events</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">chrm_pos</span><span class="p">[</span><span class="n">as_chrm_pos</span><span class="p">]),</span>
        <span class="n">np</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">chrm_signal</span><span class="p">[</span><span class="n">as_chrm_pos</span><span class="p">],</span> <span class="n">split_poss</span><span class="p">)))</span>

    <span class="k">return</span> <span class="n">cs_base_events</span>


<span class="c1">###################################</span>
<span class="c1">###### FAST5 Write Functions ######</span>
<span class="c1">###################################</span>

<span class="k">def</span> <span class="nf">prep_fast5</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">,</span> <span class="n">in_place</span><span class="p">,</span>
               <span class="n">bc_grp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">return_fp</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Prepare a read for re-squiggle processing (This deletes old re-squiggle info for this read)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">try_close_prep_err</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">,</span> <span class="n">err_str</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fast5_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="c1"># is_tombo_error = True</span>
        <span class="k">return</span> <span class="n">err_str</span><span class="p">,</span> <span class="n">fast5_fn</span><span class="p">,</span> <span class="kc">True</span>

    <span class="c1"># several checks to prepare the FAST5 file for correction before</span>
    <span class="c1"># processing to save compute</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">in_place</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="s1">&#39;Not currently implementing new hdf5 file writing&#39;</span><span class="p">,</span>
                <span class="n">fast5_fn</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
    <span class="c1"># check that the file is writeable before trying to correct</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">W_OK</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;FAST5 file is not writable&#39;</span><span class="p">,</span> <span class="n">fast5_fn</span><span class="p">,</span> <span class="kc">True</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># create group to store data</span>
        <span class="n">fast5_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_fn</span><span class="p">,</span> <span class="s1">&#39;r+&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">analyses_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">try_close_prep_err</span><span class="p">(</span>
                <span class="n">fast5_data</span><span class="p">,</span>
                <span class="s1">&#39;Base calls not found in FAST5 (see `tombo preprocess`)&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># check that the requested basecalls group exsists</span>
            <span class="k">if</span> <span class="n">bc_grp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">analyses_grp</span><span class="p">[</span><span class="n">bc_grp</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">try_close_prep_err</span><span class="p">(</span>
                <span class="n">fast5_data</span><span class="p">,</span>
                <span class="s1">&#39;Base calls not found in FAST5 (see `tombo preprocess`)&#39;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">corr_grp_ptr</span> <span class="o">=</span> <span class="n">analyses_grp</span><span class="p">[</span><span class="n">corr_grp</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">try_close_prep_err</span><span class="p">(</span>
                    <span class="n">fast5_data</span><span class="p">,</span> <span class="s2">&quot;Tombo data exists in [--corrected-group] &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;and [--overwrite] is not set&quot;</span><span class="p">)</span>
            <span class="k">del</span> <span class="n">analyses_grp</span><span class="p">[</span><span class="n">corr_grp</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="c1"># if the corr_grp isn&#39;t there we will write it now, but</span>
            <span class="c1"># it&#39;s more efficient to try than to check if the slot is there</span>
            <span class="k">pass</span>

        <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">analyses_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">corr_grp</span><span class="p">)</span>
        <span class="n">corr_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;tombo_version&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">TOMBO_VERSION</span>
        <span class="n">corr_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;basecall_group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bc_grp</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;Error opening or writing to fast5 file&#39;</span><span class="p">,</span> <span class="n">fast5_fn</span><span class="p">,</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">return_fp</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">fast5_data</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">fast5_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;Error closing fast5 file&#39;</span><span class="p">,</span> <span class="n">fast5_fn</span><span class="p">,</span> <span class="kc">True</span>

    <span class="k">return</span>

<span class="k">def</span> <span class="nf">write_error_status</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">corr_grp</span><span class="p">,</span> <span class="n">bc_subgrp</span><span class="p">,</span> <span class="n">error_text</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Write error message for a read into the FAST5 file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s1">&#39;r+&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fast5_data</span><span class="p">:</span>
        <span class="n">analysis_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">]</span>
        <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">analysis_grp</span><span class="p">[</span><span class="n">corr_grp</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">bc_subgrp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># add subgroup matching subgroup from original basecalls</span>
            <span class="n">corr_subgrp</span> <span class="o">=</span> <span class="n">corr_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">bc_subgrp</span><span class="p">)</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">error_text</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">corr_grp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">error_text</span>

    <span class="k">return</span>

<span class="k">def</span> <span class="nf">write_new_fast5_group</span><span class="p">(</span>
        <span class="n">fast5_data</span><span class="p">,</span> <span class="n">corr_grp_slot</span><span class="p">,</span> <span class="n">rsqgl_res</span><span class="p">,</span> <span class="n">norm_type</span><span class="p">,</span> <span class="n">compute_sd</span><span class="p">,</span>
        <span class="n">alignVals</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">old_segs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rna</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Write new fast5 group with re-squiggle data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># compute event data before accessing fast5 file</span>
        <span class="k">if</span> <span class="n">compute_sd</span><span class="p">:</span>
            <span class="n">norm_means</span><span class="p">,</span> <span class="n">norm_stds</span> <span class="o">=</span> <span class="n">c_new_mean_stds</span><span class="p">(</span>
                <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">raw_signal</span><span class="p">,</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">segs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">norm_means</span> <span class="o">=</span> <span class="n">c_new_means</span><span class="p">(</span><span class="n">rsqgl_res</span><span class="o">.</span><span class="n">raw_signal</span><span class="p">,</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">segs</span><span class="p">)</span>
            <span class="n">norm_stds</span> <span class="o">=</span> <span class="n">repeat</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">NAN</span><span class="p">)</span>

        <span class="c1"># had to shift to names formats numpy array specification due to</span>
        <span class="c1"># python2 numpy unicode issues. See discussion here:</span>
        <span class="c1"># https://github.com/numpy/numpy/issues/2407</span>
        <span class="n">event_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">norm_means</span><span class="p">,</span> <span class="n">norm_stds</span><span class="p">,</span>
                     <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">segs</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">rsqgl_res</span><span class="o">.</span><span class="n">segs</span><span class="p">),</span>
                     <span class="nb">list</span><span class="p">(</span><span class="n">rsqgl_res</span><span class="o">.</span><span class="n">genome_seq</span><span class="p">))),</span>
            <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;norm_mean&#39;</span><span class="p">),</span> <span class="s1">&#39;f8&#39;</span><span class="p">),</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;norm_stdev&#39;</span><span class="p">),</span> <span class="s1">&#39;f8&#39;</span><span class="p">),</span>
                   <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;start&#39;</span><span class="p">),</span> <span class="s1">&#39;u4&#39;</span><span class="p">),</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;length&#39;</span><span class="p">),</span> <span class="s1">&#39;u4&#39;</span><span class="p">),</span>
                   <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;base&#39;</span><span class="p">),</span> <span class="s1">&#39;S1&#39;</span><span class="p">)])</span>

        <span class="k">if</span> <span class="n">alignVals</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">r_align_vals</span><span class="p">,</span> <span class="n">g_align_vals</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">alignVals</span><span class="p">)</span>
            <span class="n">np_read_align</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">chararray</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignVals</span><span class="p">))</span>
            <span class="n">np_read_align</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">r_align_vals</span>
            <span class="n">np_genome_align</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">chararray</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignVals</span><span class="p">))</span>
            <span class="n">np_genome_align</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">g_align_vals</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span><span class="s1">&#39;Error computing new events&#39;</span><span class="p">)</span>

    <span class="n">do_close</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">,</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fast5_data</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">fast5_data</span><span class="p">,</span> <span class="s1">&#39;r+&#39;</span><span class="p">)</span>
            <span class="n">do_close</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span>
                <span class="s1">&#39;Error opening file for new group writing. This should &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;have been caught during the alignment phase. Check that &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;there are no other tombo processes or processes &#39;</span> <span class="o">+</span>
                <span class="s1">&#39;accessing these HDF5 files running simultaneously.&#39;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">analysis_grp</span> <span class="o">=</span> <span class="n">fast5_data</span><span class="p">[</span><span class="s1">&#39;/Analyses&#39;</span><span class="p">]</span>
        <span class="n">corr_grp</span> <span class="o">=</span> <span class="n">analysis_grp</span><span class="p">[</span><span class="n">corr_grp_slot</span><span class="p">]</span>
        <span class="c1"># add subgroup matching subgroup from original basecalls</span>
        <span class="n">corr_subgrp</span> <span class="o">=</span> <span class="n">corr_grp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">rsqgl_res</span><span class="o">.</span><span class="n">align_info</span><span class="o">.</span><span class="n">Subgroup</span><span class="p">)</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;success&#39;</span>
        <span class="c1"># TODO change to rev_sig</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna</span>
        <span class="k">if</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">sig_match_score</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;signal_match_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">sig_match_score</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;shift&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">scale_values</span><span class="o">.</span><span class="n">shift</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;scale&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">scale_values</span><span class="o">.</span><span class="n">scale</span>
        <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;norm_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">norm_type</span>
        <span class="k">if</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">scale_values</span><span class="o">.</span><span class="n">lower_lim</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;lower_lim&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">scale_values</span><span class="o">.</span><span class="n">lower_lim</span>
        <span class="k">if</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">scale_values</span><span class="o">.</span><span class="n">upper_lim</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;upper_lim&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">scale_values</span><span class="o">.</span><span class="n">upper_lim</span>
        <span class="k">if</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">scale_values</span><span class="o">.</span><span class="n">outlier_thresh</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;outlier_threshold&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">scale_values</span><span class="o">.</span><span class="n">outlier_thresh</span>

        <span class="c1"># store alignment statistics</span>
        <span class="n">corr_alignment</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="s1">&#39;Alignment&#39;</span><span class="p">)</span>
        <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;mapped_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">genome_loc</span><span class="o">.</span><span class="n">Start</span>
        <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
            <span class="s1">&#39;mapped_end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">genome_loc</span><span class="o">.</span><span class="n">Start</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">rsqgl_res</span><span class="o">.</span><span class="n">segs</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
            <span class="s1">&#39;mapped_strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">genome_loc</span><span class="o">.</span><span class="n">Strand</span>
        <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;mapped_chrom&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">genome_loc</span><span class="o">.</span><span class="n">Chrom</span>

        <span class="k">if</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">align_info</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;clipped_bases_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">align_info</span><span class="o">.</span><span class="n">ClipStart</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;clipped_bases_end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">align_info</span><span class="o">.</span><span class="n">ClipEnd</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;num_insertions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">align_info</span><span class="o">.</span><span class="n">Insertions</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;num_deletions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">align_info</span><span class="o">.</span><span class="n">Deletions</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;num_matches&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">align_info</span><span class="o">.</span><span class="n">Matches</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
                <span class="s1">&#39;num_mismatches&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">align_info</span><span class="o">.</span><span class="n">Mismatches</span>

        <span class="k">if</span> <span class="n">alignVals</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
                <span class="s1">&#39;read_alignment&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">np_read_align</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
                <span class="s1">&#39;genome_alignment&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">np_genome_align</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">old_segs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># store old segmentation in order to plot &quot;correction process&quot;</span>
            <span class="n">corr_alignment</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
                <span class="s1">&#39;read_segments&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">old_segs</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>

        <span class="c1"># Add Events to data frame with event means, SDs and lengths</span>
        <span class="n">corr_events</span> <span class="o">=</span> <span class="n">corr_subgrp</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span>
            <span class="s1">&#39;Events&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">event_data</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
        <span class="n">corr_events</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span>
            <span class="s1">&#39;read_start_rel_to_raw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rsqgl_res</span><span class="o">.</span><span class="n">read_start_rel_to_raw</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span>
            <span class="s1">&#39;Error writing resquiggle information back into fast5 file.&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">do_close</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fast5_data</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">TomboError</span><span class="p">(</span>
                <span class="s1">&#39;Error closing fast5 file after writing resquiggle information.&#39;</span><span class="p">)</span>

    <span class="k">return</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;This is a module. See commands with `tombo -h`&#39;</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-18, Oxford Nanopore Technologies

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script async="async" type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    

  

  <script type="text/javascript" src="../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>